

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>command-line interface &mdash; swibrid 2024 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=6fefd858"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="understanding output" href="output.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            swibrid
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="quickstart.html">quick start guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="configure.html">configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">understanding output</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">command-line interface</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#swibrid">swibrid</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-analyze-clustering">swibrid analyze_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-collect-results">swibrid collect_results</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-combine-replicates">swibrid combine_replicates</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-construct-linkage">swibrid construct_linkage</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-construct-msa">swibrid construct_msa</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-create-bed">swibrid create_bed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-demultiplex">swibrid demultiplex</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-downsample">swibrid downsample</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-downsample-clustering">swibrid downsample_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-find-clusters">swibrid find_clusters</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-find-rearrangements">swibrid find_rearrangements</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-find-variants">swibrid find_variants</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-alignment-pars">swibrid get_alignment_pars</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-annotation">swibrid get_annotation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-breakpoint-stats">swibrid get_breakpoint_stats</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-gaps">swibrid get_gaps</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-summary">swibrid get_summary</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-switch-homology">swibrid get_switch_homology</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-switch-motifs">swibrid get_switch_motifs</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-synthetic-reads">swibrid get_synthetic_reads</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-get-unique-clones-bed">swibrid get_unique_clones_bed</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-plot-clustering">swibrid plot_clustering</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-plot-demux-report">swibrid plot_demux_report</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-process-alignments">swibrid process_alignments</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-run">swibrid run</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-setup">swibrid setup</a></li>
<li class="toctree-l3"><a class="reference internal" href="#swibrid-test">swibrid test</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">swibrid</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">command-line interface</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/cli.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="command-line-interface">
<h1>command-line interface<a class="headerlink" href="#command-line-interface" title="Link to this heading"></a></h1>
<section id="swibrid">
<h2>swibrid<a class="headerlink" href="#swibrid" title="Link to this heading"></a></h2>
<p>Command line interface for swibrid.</p>
<p>This is the main program entry point for the <code class="docutils literal notranslate"><span class="pre">swibrid</span></code> executable and its sub commands.</p>
<p>main commands:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">setup</span></code>                  copy config and snakefile to current directory</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">run</span></code>                    run the entire pipeline</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">test</span></code>                   perform a test run on synthetic reads</p></li>
</ul>
<p>subcommands to run pipeline steps individually:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">demultiplex</span></code>            demultiplex minION run</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">process_alignments</span></code>     process and filter alignment output</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_alignment_pars</span></code>     get alignment parameters (mismatch + gap rates) from align output</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">create_bed</span></code>             create bed files and summary tables for insert-containing reads</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">construct_msa</span></code>          construct a (pseudo) MSA</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_gaps</span></code>               find gaps / breakpoints in MSA</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">construct_linkage</span></code>      construct hierarchical clustering linkage</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">find_clusters</span></code>          get read clustering from linkage</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">find_variants</span></code>          variant calling</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">find_rearrangements</span></code>    detect rearrangements (structural variants)</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">plot_clustering</span></code>        plot the read clustering</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_breakpoint_stats</span></code>   get breakpoint histograms and stats</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_switch_homology</span></code>    get switch sequence homology</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_switch_motifs</span></code>      get switch sequence motifs</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">analyze_clustering</span></code>     analyze clustering results</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">downsample_clustering</span></code>  downsample and analyze clustering</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_summary</span></code>            get sample summary stats and plot</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">collect_results</span></code>        collect results for multiple samples</p></li>
</ul>
<p>additional subcommands:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_annotation</span></code>         prepare gene annotation</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_unique_clones_bed</span></code>  get bed file with unique clones</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">get_synthetic_reads</span></code>    create synthetic reads from bed file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">plot_demux_report</span></code>      make a graphical summary of demultiplexing output</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">combine_replicates</span></code>     combine (aligned &amp; processed) reads from replicates</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">downsample</span></code>             downsample (aligned &amp; processed) reads from a sample</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid [-h] [-v] [--version]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-h">
<span id="cmdoption-swibrid-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-v">
<span id="cmdoption-swibrid-verbose"></span><span class="sig-name descname"><span class="pre">-v</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--verbose</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-v" title="Link to this definition"></a></dt>
<dd><p>Increase verbosity.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-version">
<span class="sig-name descname"><span class="pre">--version</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-version" title="Link to this definition"></a></dt>
<dd><p>show program’s version number and exit</p>
</dd></dl>

<section id="swibrid-analyze-clustering">
<h3>swibrid analyze_clustering<a class="headerlink" href="#swibrid-analyze-clustering" title="Link to this heading"></a></h3>
<p>analyze clustering: given the clustering from <code class="docutils literal notranslate"><span class="pre">find_clusters</span></code> as input,
various aggregate statistics are generated per cluster.
output table contains the following columns:</p>
<ul class="simple">
<li><p>cluster size (# of reads)</p></li>
<li><p>cluster isotype (most frequent)</p></li>
<li><p>average read fraction mapped per cluster</p></li>
<li><p>average read fraction mapped to same region</p></li>
<li><p>average read fraction ignored</p></li>
<li><p>consensus sequence</p></li>
<li><p>length of consensus</p></li>
<li><p>GC content of consensus</p></li>
<li><p>spread of breakpoints (number of bases with intermediate coverage)</p></li>
<li><p>size of inversions</p></li>
<li><p>size of duplications</p></li>
<li><p>frequency of inserts in cluster</p></li>
<li><p>fractional overlap of inserts in cluster</p></li>
<li><p>fractional overlap of insert-associated breakpoints</p></li>
<li><p>average insert length</p></li>
<li><p>average length of gaps around inserts</p></li>
<li><p>average isotype associated to inserts</p></li>
<li><p>number of homologous nucleotides around insert-associated breakpoints</p></li>
<li><p>number of homologous nucleotides around switch breakpoints</p></li>
<li><p>number of untemplated nucleotides around insert-associated breakpoints</p></li>
<li><p>number of untemplated nucleotides around switch breakpoints</p></li>
<li><p>length of different switch regions covered in cluster</p></li>
<li><p>cluster size adjusted for PCR length and GC bias</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid analyze_clustering [-h] [--clustering CLUSTERING] [--msa MSA]</span>
<span class="go">                                  [-o OUTPUT] [--gaps GAPS]</span>
<span class="go">                                  [--max_gap MAX_GAP]</span>
<span class="go">                                  [--max_realignment_gap MAX_REALIGNMENT_GAP]</span>
<span class="go">                                  [--switch_coords SWITCH_COORDS]</span>
<span class="go">                                  [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                                  [--inserts INSERTS]</span>
<span class="go">                                  [--realignments REALIGNMENTS]</span>
<span class="go">                                  [--adjust_size]</span>
<span class="go">                                  [--isotype_extra ISOTYPE_EXTRA]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-h">
<span id="cmdoption-swibrid-analyze_clustering-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-clustering">
<span class="sig-name descname"><span class="pre">--clustering</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-clustering" title="Link to this definition"></a></dt>
<dd><p>required: find_clusters output file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-msa" title="Link to this definition"></a></dt>
<dd><p>required: file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-o">
<span id="cmdoption-swibrid-analyze_clustering-output"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-o" title="Link to this definition"></a></dt>
<dd><p>required: output table</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-gaps">
<span class="sig-name descname"><span class="pre">--gaps</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gaps&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-gaps" title="Link to this definition"></a></dt>
<dd><p>required: gap distribution</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-max_gap">
<span class="sig-name descname"><span class="pre">--max_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-max_gap" title="Link to this definition"></a></dt>
<dd><p>max gap size to ignore [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-max_realignment_gap">
<span class="sig-name descname"><span class="pre">--max_realignment_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_realignment_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-max_realignment_gap" title="Link to this definition"></a></dt>
<dd><p>max gap size to count positions for realignment [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-inserts">
<span class="sig-name descname"><span class="pre">--inserts</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inserts&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-inserts" title="Link to this definition"></a></dt>
<dd><p>results.tsv file with inserts</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-realignments">
<span class="sig-name descname"><span class="pre">--realignments</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;realignments&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-realignments" title="Link to this definition"></a></dt>
<dd><p>file with breakpoint realignments</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-adjust_size">
<span class="sig-name descname"><span class="pre">--adjust_size</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-adjust_size" title="Link to this definition"></a></dt>
<dd><p>calculate cluster size adjusted for fragment length and QC content</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-analyze_clustering-isotype_extra">
<span class="sig-name descname"><span class="pre">--isotype_extra</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;isotype_extra&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-analyze_clustering-isotype_extra" title="Link to this definition"></a></dt>
<dd><p>extra space added to define isotypes [500]</p>
</dd></dl>

</section>
<section id="swibrid-collect-results">
<h3>swibrid collect_results<a class="headerlink" href="#swibrid-collect-results" title="Link to this heading"></a></h3>
<p>collect stats from different samples: this will collect all summary files for the listed samples and produce one big table.
optionally, this will also collect all inserts and produce a multi-sheet excel file.
clustering analysis results can also be collected into an excel file.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid collect_results [-h] --samples SAMPLES --sample_stats</span>
<span class="go">                               SAMPLE_STATS [--summary_path SUMMARY_PATH]</span>
<span class="go">                               [--inserts_path INSERTS_PATH]</span>
<span class="go">                               [--cluster_analysis_path CLUSTER_ANALYSIS_PATH]</span>
<span class="go">                               [--cluster_stats CLUSTER_STATS]</span>
<span class="go">                               [--inserts INSERTS] [--clean_column_names]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-h">
<span id="cmdoption-swibrid-collect_results-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-samples">
<span class="sig-name descname"><span class="pre">--samples</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;samples&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-samples" title="Link to this definition"></a></dt>
<dd><p>required: comma-separated list of samples</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-sample_stats">
<span class="sig-name descname"><span class="pre">--sample_stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample_stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-sample_stats" title="Link to this definition"></a></dt>
<dd><p>required: output file with sample_stats</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-summary_path">
<span class="sig-name descname"><span class="pre">--summary_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;summary_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-summary_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for summary files [pipeline/{sample}/{sample}_summary.csv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-inserts_path">
<span class="sig-name descname"><span class="pre">--inserts_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inserts_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-inserts_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for insert tables [pipeline/{sample}/{sample}_inserts.tsv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-cluster_analysis_path">
<span class="sig-name descname"><span class="pre">--cluster_analysis_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cluster_analysis_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-cluster_analysis_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for cluster_analysis files [pipeline/{sample}/{sample}_cluster_analysis.csv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-cluster_stats">
<span class="sig-name descname"><span class="pre">--cluster_stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cluster_stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-cluster_stats" title="Link to this definition"></a></dt>
<dd><p>output file with cluster stats</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-inserts">
<span class="sig-name descname"><span class="pre">--inserts</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inserts&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-inserts" title="Link to this definition"></a></dt>
<dd><p>output file with inserts</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-collect_results-clean_column_names">
<span class="sig-name descname"><span class="pre">--clean_column_names</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-collect_results-clean_column_names" title="Link to this definition"></a></dt>
<dd><p>re-name columns to make features a bit more explicit</p>
</dd></dl>

</section>
<section id="swibrid-combine-replicates">
<h3>swibrid combine_replicates<a class="headerlink" href="#swibrid-combine-replicates" title="Link to this heading"></a></h3>
<p>combine (aligned + processed) reads from replicates</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid combine_replicates [-h] [-s SAMPLES] [-c COMBINED]</span>
<span class="go">                                  [--nmax NMAX] [--aligner ALIGNER]</span>
<span class="go">                                  [--msa_path MSA_PATH]</span>
<span class="go">                                  [--msa_csv_path MSA_CSV_PATH]</span>
<span class="go">                                  [--info_path INFO_PATH]</span>
<span class="go">                                  [--process_path PROCESS_PATH]</span>
<span class="go">                                  [--process_stats_path PROCESS_STATS_PATH]</span>
<span class="go">                                  [--alignment_pars_path ALIGNMENT_PARS_PATH]</span>
<span class="go">                                  [--aligned_seqs_path ALIGNED_SEQS_PATH]</span>
<span class="go">                                  [--breakpoint_alignments_path BREAKPOINT_ALIGNMENTS_PATH]</span>
<span class="go">                                  [--inserts_tsv_path INSERTS_TSV_PATH]</span>
<span class="go">                                  [--bed_path BED_PATH]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-h">
<span id="cmdoption-swibrid-combine_replicates-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-s">
<span id="cmdoption-swibrid-combine_replicates-samples"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;samples&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--samples</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;samples&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-s" title="Link to this definition"></a></dt>
<dd><p>comma-separated list of samples to combine</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-c">
<span id="cmdoption-swibrid-combine_replicates-combined"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;combined&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--combined</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;combined&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-c" title="Link to this definition"></a></dt>
<dd><p>combined sample name</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-nmax">
<span class="sig-name descname"><span class="pre">--nmax</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nmax&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-nmax" title="Link to this definition"></a></dt>
<dd><p>use only nmax reads [50000]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-aligner">
<span class="sig-name descname"><span class="pre">--aligner</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;aligner&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-aligner" title="Link to this definition"></a></dt>
<dd><p>alignment algorithm used [last]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-msa_path">
<span class="sig-name descname"><span class="pre">--msa_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-msa_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for msa files [“pipeline/{sample}/{sample}_msa.npz”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-msa_csv_path">
<span class="sig-name descname"><span class="pre">--msa_csv_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa_csv_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-msa_csv_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for msa csv files [“pipeline/{sample}/{sample}_msa.csv”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-info_path">
<span class="sig-name descname"><span class="pre">--info_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-info_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for info csv files [“input/{sample}_info.csv”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-process_path">
<span class="sig-name descname"><span class="pre">--process_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;process_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-process_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for processing output [“pipeline/{sample}/{sample}_processed.out”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-process_stats_path">
<span class="sig-name descname"><span class="pre">--process_stats_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;process_stats_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-process_stats_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for process stats csv [“pipeline/{sample}/{sample}_process_stats.csv”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-alignment_pars_path">
<span class="sig-name descname"><span class="pre">--alignment_pars_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;alignment_pars_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-alignment_pars_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for alignment pars [“pipeline/{sample}/{sample}_{aligner}_pars.npz”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-aligned_seqs_path">
<span class="sig-name descname"><span class="pre">--aligned_seqs_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;aligned_seqs_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-aligned_seqs_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for aligned sequences [“pipeline/{sample}/{sample}_aligned.fasta.gz</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-breakpoint_alignments_path">
<span class="sig-name descname"><span class="pre">--breakpoint_alignments_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;breakpoint_alignments_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-breakpoint_alignments_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for breakpoint alignments [“pipeline/{sample}/{sample}_breakpoint_alignments.csv”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-inserts_tsv_path">
<span class="sig-name descname"><span class="pre">--inserts_tsv_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inserts_tsv_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-inserts_tsv_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for inserts tsv [“pipeline/{sample}/{sample}_inserts.tsv”]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-combine_replicates-bed_path">
<span class="sig-name descname"><span class="pre">--bed_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-combine_replicates-bed_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for bed [“pipeline/{sample}/{sample}.bed”]</p>
</dd></dl>

</section>
<section id="swibrid-construct-linkage">
<h3>swibrid construct_linkage<a class="headerlink" href="#swibrid-construct-linkage" title="Link to this heading"></a></h3>
<p>construct hiearchical agglomerative clustering from MSA:
from the input MSA, only coverage information will be used and gaps smaller than <cite>max_gap</cite> will be removed.
by default, <cite>fastcluster</cite> is used with cosine metric and average linkage.
output is a npz file containing the linkage matrix</p>
<p>EXPERIMENTAL: a sparse version building on pynndescent and gbbs agglomerative clustering, requires several libraries to be installed</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid construct_linkage [-h] [--msa MSA] [--linkage LINKAGE]</span>
<span class="go">                                 [--gaps GAPS] [--max_gap MAX_GAP]</span>
<span class="go">                                 [--metric METRIC] [--method METHOD]</span>
<span class="go">                                 [--nmax NMAX] [--ignore_unused_positions]</span>
<span class="go">                                 [--use_sparse] [--distances DISTANCES]</span>
<span class="go">                                 [--n_neighbors N_NEIGHBORS]</span>
<span class="go">                                 [--n_threads N_THREADS] [--n_backup N_BACKUP]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-h">
<span id="cmdoption-swibrid-construct_linkage-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-msa" title="Link to this definition"></a></dt>
<dd><p>required: file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-linkage">
<span class="sig-name descname"><span class="pre">--linkage</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;linkage&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-linkage" title="Link to this definition"></a></dt>
<dd><p>required: output file contains linkage</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-gaps">
<span class="sig-name descname"><span class="pre">--gaps</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gaps&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-gaps" title="Link to this definition"></a></dt>
<dd><p>output of get_gaps</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-max_gap">
<span class="sig-name descname"><span class="pre">--max_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-max_gap" title="Link to this definition"></a></dt>
<dd><p>max gap size to ignore [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-metric">
<span class="sig-name descname"><span class="pre">--metric</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;metric&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-metric" title="Link to this definition"></a></dt>
<dd><p>clustering metric [cosine]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-method">
<span class="sig-name descname"><span class="pre">--method</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;method&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-method" title="Link to this definition"></a></dt>
<dd><p>clustering method for hierarchical clustering [average]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-nmax">
<span class="sig-name descname"><span class="pre">--nmax</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nmax&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-nmax" title="Link to this definition"></a></dt>
<dd><p>use only nmax reads</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-ignore_unused_positions">
<span class="sig-name descname"><span class="pre">--ignore_unused_positions</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-ignore_unused_positions" title="Link to this definition"></a></dt>
<dd><p>ignore positions that have no coverage</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-use_sparse">
<span class="sig-name descname"><span class="pre">--use_sparse</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-use_sparse" title="Link to this definition"></a></dt>
<dd><p>EXPERIMENTAL: use sparse nearest-neighbor clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-distances">
<span class="sig-name descname"><span class="pre">--distances</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;distances&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-distances" title="Link to this definition"></a></dt>
<dd><p>pre-computed sparse distance matrix</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-n_neighbors">
<span class="sig-name descname"><span class="pre">--n_neighbors</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n_neighbors&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-n_neighbors" title="Link to this definition"></a></dt>
<dd><p># of nearest neighbors for adjacency graph [100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-n_threads">
<span class="sig-name descname"><span class="pre">--n_threads</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n_threads&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-n_threads" title="Link to this definition"></a></dt>
<dd><p># of threads [1]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_linkage-n_backup">
<span class="sig-name descname"><span class="pre">--n_backup</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n_backup&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_linkage-n_backup" title="Link to this definition"></a></dt>
<dd><p>save full distances for these points in addition to the nearest neighbors [0]</p>
</dd></dl>

</section>
<section id="swibrid-construct-msa">
<h3>swibrid construct_msa<a class="headerlink" href="#swibrid-construct-msa" title="Link to this heading"></a></h3>
<p>construct (pseudo) MSA from processed alignments:
this script will take aligned segments and sequences from process_alignments and construct a MSA.
the MSA is stored as a sparse matrix of n_reads x n_positions, where the positions run over the concatenation of individual switch regions specified in a bed file.
matrix values m code for coverage and nucleotide identity:</p>
<ul class="simple">
<li><p>m % 10 gives nucleotide identity with A=1, C=2, G=3, T=4, gaps are zeros</p></li>
<li><p>m // 10 gives coverage of genomic regions by one read (1x, 2x, …)  and indicates tandem duplications; if <code class="docutils literal notranslate"><span class="pre">--use_orientation</span></code> is set, inversions are also considered and associated coverage values are negative</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid construct_msa [-h] [--coords COORDS] [--sequences SEQUENCES]</span>
<span class="go">                             [--msa MSA] [--out OUT]</span>
<span class="go">                             [--switch_coords SWITCH_COORDS]</span>
<span class="go">                             [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                             [--nmax NMAX] [--use_orientation]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-h">
<span id="cmdoption-swibrid-construct_msa-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-coords">
<span class="sig-name descname"><span class="pre">--coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-coords" title="Link to this definition"></a></dt>
<dd><p>required: process_alignments coordinate output</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-sequences">
<span class="sig-name descname"><span class="pre">--sequences</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sequences&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-sequences" title="Link to this definition"></a></dt>
<dd><p>required: process_alignments sequence output</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-msa" title="Link to this definition"></a></dt>
<dd><p>required: output file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-out">
<span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-out" title="Link to this definition"></a></dt>
<dd><p>required: output file with read info</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>required: bed file with coordinates of individual switch regions</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-nmax">
<span class="sig-name descname"><span class="pre">--nmax</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nmax&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-nmax" title="Link to this definition"></a></dt>
<dd><p>use only nmax reads</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-construct_msa-use_orientation">
<span class="sig-name descname"><span class="pre">--use_orientation</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-construct_msa-use_orientation" title="Link to this definition"></a></dt>
<dd><p>include alignment block orientation</p>
</dd></dl>

</section>
<section id="swibrid-create-bed">
<h3>swibrid create_bed<a class="headerlink" href="#swibrid-create-bed" title="Link to this heading"></a></h3>
<p>create bed files for all reads and summary table for insert-containing reads, including read sequence with insert part highlighted in uppercase</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid create_bed [-h] [--processed_alignments PROCESSED_ALIGNMENTS]</span>
<span class="go">                          [--bed BED] [--outfile OUTFILE]</span>
<span class="go">                          [--raw_reads RAW_READS] [--paired_end_mode]</span>
<span class="go">                          [--switch_coords SWITCH_COORDS]</span>
<span class="go">                          [--annotation [ANNOTATION]]</span>
<span class="go">                          [--switch_annotation SWITCH_ANNOTATION]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-h">
<span id="cmdoption-swibrid-create_bed-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-processed_alignments">
<span class="sig-name descname"><span class="pre">--processed_alignments</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;processed_alignments&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-processed_alignments" title="Link to this definition"></a></dt>
<dd><p>required: output of process_alignments</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-bed">
<span class="sig-name descname"><span class="pre">--bed</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-bed" title="Link to this definition"></a></dt>
<dd><p>required: bed output with insert + switch coordinates for selected reads</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-outfile">
<span class="sig-name descname"><span class="pre">--outfile</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;outfile&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-outfile" title="Link to this definition"></a></dt>
<dd><p>required: output table with insert information</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-raw_reads">
<span class="sig-name descname"><span class="pre">--raw_reads</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;raw_reads&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-raw_reads" title="Link to this definition"></a></dt>
<dd><p>fasta file with minION reads (interleaved mates for paired-end data)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-paired_end_mode">
<span class="sig-name descname"><span class="pre">--paired_end_mode</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-paired_end_mode" title="Link to this definition"></a></dt>
<dd><p>EXPERIMENTAL: use paired-end mode (use interleaved mates as input)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-annotation">
<span class="sig-name descname"><span class="pre">--annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with gene annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-create_bed-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-create_bed-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

</section>
<section id="swibrid-demultiplex">
<h3>swibrid demultiplex<a class="headerlink" href="#swibrid-demultiplex" title="Link to this heading"></a></h3>
<p>demultiplex reads using output of BLAST against primers and barcodes</p>
<p>BLAST output is created with <code class="docutils literal notranslate"><span class="pre">blastn</span> <span class="pre">-db</span> <span class="pre">{barcodes_primers}</span> <span class="pre">-query</span> <span class="pre">{query}</span> <span class="pre">-task</span> <span class="pre">blastn-short</span> <span class="pre">-max_target_seqs</span> <span class="pre">50</span> <span class="pre">-outfmt</span> <span class="pre">&quot;6</span> <span class="pre">saccver</span> <span class="pre">qaccver</span> <span class="pre">slen</span> <span class="pre">qlen</span> <span class="pre">pident</span> <span class="pre">length</span> <span class="pre">qstart</span> <span class="pre">qend</span> <span class="pre">evalue&quot;</span> <span class="pre">-gapopen</span> <span class="pre">5</span> <span class="pre">-gapextend</span> <span class="pre">2</span> <span class="pre">-reward</span> <span class="pre">3</span> <span class="pre">-penalty</span> <span class="pre">-4</span> <span class="pre">-evalue</span> <span class="pre">1</span> <span class="pre">-num_threads</span> <span class="pre">1</span> <span class="pre">-perc_identity</span> <span class="pre">50</span></code> where {barcodes_primers} is a database containing barcode and primer sequences and {query} is the input fasta file</p>
<p>optional (but recommended) input is a sample sheet (tab delimited, no header or row names, 1st column barcode, 2nd column sample name)</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid demultiplex [-h] -i INPUT -b BLAST -o OUTDIR [-c CUTOFF]</span>
<span class="go">                           [-d DIST] [-m MIN_READS] [-f FIGURE] [-r REPORT]</span>
<span class="go">                           [-s SAMPLE_SHEET] [--collapse] [--split_reads]</span>
<span class="go">                           [--max_split_dist MAX_SPLIT_DIST] [--nmax NMAX]</span>
<span class="go">                           [--add_umi] [--umi_regex UMI_REGEX]</span>
<span class="go">                           [--write_info_chunks]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-h">
<span id="cmdoption-swibrid-demultiplex-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-i">
<span id="cmdoption-swibrid-demultiplex-input"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--input</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-i" title="Link to this definition"></a></dt>
<dd><p>required: input fastq(.gz)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-b">
<span id="cmdoption-swibrid-demultiplex-blast"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;blast&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--blast</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;blast&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-b" title="Link to this definition"></a></dt>
<dd><p>required: BLAST output file (or stdin)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-o">
<span id="cmdoption-swibrid-demultiplex-outdir"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;outdir&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--outdir</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;outdir&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-o" title="Link to this definition"></a></dt>
<dd><p>required: output directory</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-c">
<span id="cmdoption-swibrid-demultiplex-cutoff"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cutoff&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-c" title="Link to this definition"></a></dt>
<dd><p>% identify cutoff [70.0]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-d">
<span id="cmdoption-swibrid-demultiplex-dist"></span><span class="sig-name descname"><span class="pre">-d</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;dist&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--dist</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;dist&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-d" title="Link to this definition"></a></dt>
<dd><p>max distance of match to end of sequence [100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-m">
<span id="cmdoption-swibrid-demultiplex-min_reads"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_reads&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--min_reads</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_reads&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-m" title="Link to this definition"></a></dt>
<dd><p>min # of reads required to create separate file if no sample sheet is given [1000]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-f">
<span id="cmdoption-swibrid-demultiplex-figure"></span><span class="sig-name descname"><span class="pre">-f</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--figure</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-f" title="Link to this definition"></a></dt>
<dd><p>summary figure</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-r">
<span id="cmdoption-swibrid-demultiplex-report"></span><span class="sig-name descname"><span class="pre">-r</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;report&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--report</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;report&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-r" title="Link to this definition"></a></dt>
<dd><p>summary report</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-s">
<span id="cmdoption-swibrid-demultiplex-sample_sheet"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample_sheet&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--sample_sheet</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample_sheet&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-s" title="Link to this definition"></a></dt>
<dd><p>sample sheet (barcode &lt;tab&gt; sample_name, no header)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-collapse">
<span class="sig-name descname"><span class="pre">--collapse</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-collapse" title="Link to this definition"></a></dt>
<dd><p>count reads regardless of whether one or both barcodes are present (for identical barcodes)? [False]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-split_reads">
<span class="sig-name descname"><span class="pre">--split_reads</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-split_reads" title="Link to this definition"></a></dt>
<dd><p>split reads with adjacent internal primer binding sites</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-max_split_dist">
<span class="sig-name descname"><span class="pre">--max_split_dist</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_split_dist&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-max_split_dist" title="Link to this definition"></a></dt>
<dd><p>maximum distance between internal primer matches to split read [200]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-nmax">
<span class="sig-name descname"><span class="pre">--nmax</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nmax&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-nmax" title="Link to this definition"></a></dt>
<dd><p>maximum number of reads</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-add_umi">
<span class="sig-name descname"><span class="pre">--add_umi</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-add_umi" title="Link to this definition"></a></dt>
<dd><p>add UMI to read name</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-umi_regex">
<span class="sig-name descname"><span class="pre">--umi_regex</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;umi_regex&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-umi_regex" title="Link to this definition"></a></dt>
<dd><p>regex to find UMI in read description [[ACGTN]{12}]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-demultiplex-write_info_chunks">
<span class="sig-name descname"><span class="pre">--write_info_chunks</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-demultiplex-write_info_chunks" title="Link to this definition"></a></dt>
<dd><p>write info files in chunks</p>
</dd></dl>

</section>
<section id="swibrid-downsample">
<h3>swibrid downsample<a class="headerlink" href="#swibrid-downsample" title="Link to this heading"></a></h3>
<p>downsample (aligned + processed) reads for a sample</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid downsample [-h] [-i INPUT] [-o OUTPUT] [-n N]</span>
<span class="go">                          [--aligner ALIGNER] [--msa_path MSA_PATH]</span>
<span class="go">                          [--msa_csv_path MSA_CSV_PATH]</span>
<span class="go">                          [--info_path INFO_PATH]</span>
<span class="go">                          [--process_path PROCESS_PATH]</span>
<span class="go">                          [--process_stats_path PROCESS_STATS_PATH]</span>
<span class="go">                          [--alignment_pars_path ALIGNMENT_PARS_PATH]</span>
<span class="go">                          [--breakpoint_alignments_path BREAKPOINT_ALIGNMENTS_PATH]</span>
<span class="go">                          [--inserts_tsv_path INSERTS_TSV_PATH]</span>
<span class="go">                          [--bed_path BED_PATH]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-h">
<span id="cmdoption-swibrid-downsample-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-downsample-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-i">
<span id="cmdoption-swibrid-downsample-input"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--input</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-i" title="Link to this definition"></a></dt>
<dd><p>input sample to downsample</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-o">
<span id="cmdoption-swibrid-downsample-output"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-o" title="Link to this definition"></a></dt>
<dd><p>downsampled sample name</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-n">
<span id="cmdoption-swibrid-downsample-n"></span><span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-n" title="Link to this definition"></a></dt>
<dd><p>use only n reads [5000]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-aligner">
<span class="sig-name descname"><span class="pre">--aligner</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;aligner&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-aligner" title="Link to this definition"></a></dt>
<dd><p>alignment algorithm used [last]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-msa_path">
<span class="sig-name descname"><span class="pre">--msa_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-msa_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for msa files [pipeline/{sample}/{sample}_msa.npz]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-msa_csv_path">
<span class="sig-name descname"><span class="pre">--msa_csv_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa_csv_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-msa_csv_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for msa csv files [pipeline/{sample}/{sample}_msa.csv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-info_path">
<span class="sig-name descname"><span class="pre">--info_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-info_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for info csv files [input/{sample}_info.csv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-process_path">
<span class="sig-name descname"><span class="pre">--process_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;process_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-process_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for processing output [pipeline/{sample}/{sample}_processed.out]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-process_stats_path">
<span class="sig-name descname"><span class="pre">--process_stats_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;process_stats_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-process_stats_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for process stats csv [pipeline/{sample}/{sample}_process_stats.csv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-alignment_pars_path">
<span class="sig-name descname"><span class="pre">--alignment_pars_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;alignment_pars_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-alignment_pars_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for alignment pars [pipeline/{sample}/{sample}_{aligner}_pars.npz]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-breakpoint_alignments_path">
<span class="sig-name descname"><span class="pre">--breakpoint_alignments_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;breakpoint_alignments_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-breakpoint_alignments_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for breakpoint alignments [pipeline/{sample}/{sample}_breakpoint_alignments.csv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-inserts_tsv_path">
<span class="sig-name descname"><span class="pre">--inserts_tsv_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inserts_tsv_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-inserts_tsv_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for inserts tsv [pipeline/{sample}/{sample}_inserts.tsv]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample-bed_path">
<span class="sig-name descname"><span class="pre">--bed_path</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed_path&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample-bed_path" title="Link to this definition"></a></dt>
<dd><p>path pattern for bed [pipeline/{sample}/{sample}.bed]</p>
</dd></dl>

</section>
<section id="swibrid-downsample-clustering">
<h3>swibrid downsample_clustering<a class="headerlink" href="#swibrid-downsample-clustering" title="Link to this heading"></a></h3>
<p>downsample the clustering to get more robust diversity measures:
given the input MSA (and associated gaps), clustering is repeated <cite>nreps</cite> times on a subsample of <cite>nreads</cite> reads
using <cite>fastcluster</cite>  with (by defalt) cosine metric and average linkage and a cutoff derived from cluster_stats input.
the following diversity measured are calculated:</p>
<ul class="simple">
<li><p>mean_cluster_size (as fraction of reads)</p></li>
<li><p>std_cluster_size</p></li>
<li><p>nclusters_final (number of clusters after filtering)</p></li>
<li><p>nclusters_eff (from the entropy of the cluster size distribution pre-filtering)</p></li>
<li><p>cluster_gini: gini coefficient (post-filtering)</p></li>
<li><p>cluster_entropy: entropy (post-filtering)</p></li>
<li><p>cluster_inverse_simpson: inverse simpson coefficient (post-filtering)</p></li>
<li><p>top_clone_occupancy: relative fraction of reads in biggest cluster</p></li>
<li><p>big_clones_occupancy: fraction of reads in clusters &gt; 1% occupancy</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid downsample_clustering [-h] [--msa MSA]</span>
<span class="go">                                     [--cluster_stats CLUSTER_STATS]</span>
<span class="go">                                     [-s STATS] [--nreads NREADS]</span>
<span class="go">                                     [--nreps NREPS] [--gaps GAPS]</span>
<span class="go">                                     [--max_gap MAX_GAP] [--metric METRIC]</span>
<span class="go">                                     [--method METHOD]</span>
<span class="go">                                     [--ignore_unused_positions]</span>
<span class="go">                                     [--filtering_cutoff FILTERING_CUTOFF]</span>
<span class="go">                                     [--big_clone_cutoff BIG_CLONE_CUTOFF]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-h">
<span id="cmdoption-swibrid-downsample_clustering-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-msa" title="Link to this definition"></a></dt>
<dd><p>required: file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-cluster_stats">
<span class="sig-name descname"><span class="pre">--cluster_stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cluster_stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-cluster_stats" title="Link to this definition"></a></dt>
<dd><p>required: file with statistics from original clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-s">
<span id="cmdoption-swibrid-downsample_clustering-stats"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-s" title="Link to this definition"></a></dt>
<dd><p>required: output file with clustering statistics after downsampling</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-nreads">
<span class="sig-name descname"><span class="pre">--nreads</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nreads&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-nreads" title="Link to this definition"></a></dt>
<dd><p>number of reads used [1000]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-nreps">
<span class="sig-name descname"><span class="pre">--nreps</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nreps&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-nreps" title="Link to this definition"></a></dt>
<dd><p>number of replicates [10]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-gaps">
<span class="sig-name descname"><span class="pre">--gaps</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gaps&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-gaps" title="Link to this definition"></a></dt>
<dd><p>output of get_gaps</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-max_gap">
<span class="sig-name descname"><span class="pre">--max_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-max_gap" title="Link to this definition"></a></dt>
<dd><p>max gap size to ignore [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-metric">
<span class="sig-name descname"><span class="pre">--metric</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;metric&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-metric" title="Link to this definition"></a></dt>
<dd><p>clustering metric [cosine]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-method">
<span class="sig-name descname"><span class="pre">--method</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;method&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-method" title="Link to this definition"></a></dt>
<dd><p>clustering method for hierarchical clustering [average]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-ignore_unused_positions">
<span class="sig-name descname"><span class="pre">--ignore_unused_positions</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-ignore_unused_positions" title="Link to this definition"></a></dt>
<dd><p>ignore positions that have no coverage</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-filtering_cutoff">
<span class="sig-name descname"><span class="pre">--filtering_cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;filtering_cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-filtering_cutoff" title="Link to this definition"></a></dt>
<dd><p>filter out smallest clusters, keeping at least this fraction of reads [0.95]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-downsample_clustering-big_clone_cutoff">
<span class="sig-name descname"><span class="pre">--big_clone_cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;big_clone_cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-downsample_clustering-big_clone_cutoff" title="Link to this definition"></a></dt>
<dd><p>cutoff to determine a “big” clone as fraction of clustered reads [0.01]</p>
</dd></dl>

</section>
<section id="swibrid-find-clusters">
<h3>swibrid find_clusters<a class="headerlink" href="#swibrid-find-clusters" title="Link to this heading"></a></h3>
<p>find clusters by cutting linkage dendrogram: inputs are the linkage matrix (from <code class="docutils literal notranslate"><span class="pre">construct_linkage</span></code>)
and read info (<code class="docutils literal notranslate"><span class="pre">from</span> <span class="pre">construct_msa</span></code>). output is a csv file that adds cluster identity to the read info file</p>
<p>cutoff can be fixed or determined by scanning cutoffs between cmin and cmax
and finding an inflection point (either by distance, or from the intersection of two exponential trends).
small and isolated clusters are flagged, filtered clusters contain at least 95% of reads</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid find_clusters [-h] -l LINKAGE -i INPUT -o OUTPUT [-s STATS]</span>
<span class="go">                             [-f CUTOFF] [--scanning SCANNING]</span>
<span class="go">                             [--fit_method FIT_METHOD] [-c CMIN] [-C CMAX]</span>
<span class="go">                             [-n NC] [--filtering_cutoff FILTERING_CUTOFF]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-h">
<span id="cmdoption-swibrid-find_clusters-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-l">
<span id="cmdoption-swibrid-find_clusters-linkage"></span><span class="sig-name descname"><span class="pre">-l</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;linkage&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--linkage</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;linkage&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-l" title="Link to this definition"></a></dt>
<dd><p>required: output of construct_linkage</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-i">
<span id="cmdoption-swibrid-find_clusters-input"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--input</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-i" title="Link to this definition"></a></dt>
<dd><p>required: read info (output of construct_msa)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-o">
<span id="cmdoption-swibrid-find_clusters-output"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-o" title="Link to this definition"></a></dt>
<dd><p>required: output file with clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-s">
<span id="cmdoption-swibrid-find_clusters-stats"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-s" title="Link to this definition"></a></dt>
<dd><p>file with statistics</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-f">
<span id="cmdoption-swibrid-find_clusters-fix_cutoff"></span><span class="sig-name descname"><span class="pre">-f</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cutoff&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--fix_cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-f" title="Link to this definition"></a></dt>
<dd><p>use fixed cutoff instead of data-derived</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-scanning">
<span class="sig-name descname"><span class="pre">--scanning</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;scanning&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-scanning" title="Link to this definition"></a></dt>
<dd><p>file with scanning data</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-fit_method">
<span class="sig-name descname"><span class="pre">--fit_method</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;fit_method&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-fit_method" title="Link to this definition"></a></dt>
<dd><p>cutoff determination method: “trend” or “distance” [distance]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-c">
<span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cmin&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-c" title="Link to this definition"></a></dt>
<dd><p>minimum cutoff value [0.001]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-C">
<span class="sig-name descname"><span class="pre">-C</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cmax&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-C" title="Link to this definition"></a></dt>
<dd><p>maximum cutoff value [0.100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-n">
<span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nc&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-n" title="Link to this definition"></a></dt>
<dd><p>number of cutoff values between cmin and cmax [100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_clusters-filtering_cutoff">
<span class="sig-name descname"><span class="pre">--filtering_cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;filtering_cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_clusters-filtering_cutoff" title="Link to this definition"></a></dt>
<dd><p>filter out smallest clusters, keeping at least this fraction of reads [0.95]</p>
</dd></dl>

</section>
<section id="swibrid-find-rearrangements">
<h3>swibrid find_rearrangements<a class="headerlink" href="#swibrid-find-rearrangements" title="Link to this heading"></a></h3>
<p>find rearrangements (inversions/duplications) in MSA: regions with coverage larger or smaller than 1 and without gaps bigger than 25 nt. output files:</p>
<ul class="simple">
<li><p>a .npz file with read indices, left/right positions and sizes for inversions and duplications per read</p></li>
<li><p>a bed file with consensus rearrangement coordinates (by merging individual events), type and allele frequency</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid find_rearrangements [-h] --msa MSA -o OUT -m MAT</span>
<span class="go">                                   [--switch_coords SWITCH_COORDS]</span>
<span class="go">                                   [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                                   [--max_rearrangement_gap MAX_REARRANGEMENT_GAP]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-h">
<span id="cmdoption-swibrid-find_rearrangements-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-msa" title="Link to this definition"></a></dt>
<dd><p>required: file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-o">
<span id="cmdoption-swibrid-find_rearrangements-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-o" title="Link to this definition"></a></dt>
<dd><p>required: output file (bed)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-m">
<span id="cmdoption-swibrid-find_rearrangements-mat"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;mat&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--mat</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;mat&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-m" title="Link to this definition"></a></dt>
<dd><p>required: output file (matrix)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_rearrangements-max_rearrangement_gap">
<span class="sig-name descname"><span class="pre">--max_rearrangement_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_rearrangement_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_rearrangements-max_rearrangement_gap" title="Link to this definition"></a></dt>
<dd><p>max gap within rearrangements to ignore [25]</p>
</dd></dl>

</section>
<section id="swibrid-find-variants">
<h3>swibrid find_variants<a class="headerlink" href="#swibrid-find-variants" title="Link to this heading"></a></h3>
<p>EXPERIMENTAL: find single-nucleotide variants in MSA and determine haplotypes. variable positions in the input MSA (excluding positions at or near gaps) where nucleotide distributions are different than what’s expected given mutation frequencies (estimated at the initial alignment step). potential variants are removed if</p>
<ul class="simple">
<li><p>there’s less than min_cov non-gaps</p></li>
<li><p>fewer variants than expected</p></li>
<li><p>high strand bias</p></li>
<li><p>no cluster with at least min_cluster_cov reads and allele frequency &gt; min_freq</p></li>
</ul>
<p>variants are aggregated over clusters, and the distribution across clusters is tested for evenness. variants can be annotated by dbSNP id. motif occurrences around variants are also scored. haplotypes are determined by performing a weighted NMF on a matrix of allele frequencies per cluster and variant.</p>
<p>output is a vcf-style file with genomic (1-based) and relative (0-based) coordinates, and a matrix (sparse integer array, same shape as MSA) indicating which read contains a variant at which position</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid find_variants [-h] [--msa MSA] [--clustering CLUSTERING]</span>
<span class="go">                             [--stats STATS] [--reference REFERENCE]</span>
<span class="go">                             [--pars PARS] [-o OUT] [-m MAT]</span>
<span class="go">                             [--save_all SAVE_ALL]</span>
<span class="go">                             [--switch_coords SWITCH_COORDS]</span>
<span class="go">                             [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                             [--fdr FDR] [--min_cov MIN_COV]</span>
<span class="go">                             [--gap_window_size GAP_WINDOW_SIZE]</span>
<span class="go">                             [--max_local_gap_freq MAX_LOCAL_GAP_FREQ]</span>
<span class="go">                             [--min_cluster_cov MIN_CLUSTER_COV]</span>
<span class="go">                             [--min_freq MIN_FREQ] [--nreads NREADS]</span>
<span class="go">                             [--variant_annotation [VARIANT_ANNOTATION]]</span>
<span class="go">                             [--haplotypes HAPLOTYPES] [--motifs MOTIFS]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-h">
<span id="cmdoption-swibrid-find_variants-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-msa" title="Link to this definition"></a></dt>
<dd><p>required: file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-clustering">
<span class="sig-name descname"><span class="pre">--clustering</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-clustering" title="Link to this definition"></a></dt>
<dd><p>required: clustering (find_clusters output)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-stats">
<span class="sig-name descname"><span class="pre">--stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-stats" title="Link to this definition"></a></dt>
<dd><p>required: clustering stats (find_clusters output)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-reference">
<span class="sig-name descname"><span class="pre">--reference</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;reference&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-reference" title="Link to this definition"></a></dt>
<dd><p>required: reference sequence (switch chromosome)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-pars">
<span class="sig-name descname"><span class="pre">--pars</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;pars&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-pars" title="Link to this definition"></a></dt>
<dd><p>required: alignment parameters to estimate mutation probability</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-o">
<span id="cmdoption-swibrid-find_variants-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-o" title="Link to this definition"></a></dt>
<dd><p>required: output file (text)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-m">
<span id="cmdoption-swibrid-find_variants-mat"></span><span class="sig-name descname"><span class="pre">-m</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;mat&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--mat</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;mat&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-m" title="Link to this definition"></a></dt>
<dd><p>required: output file (matrix)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-save_all">
<span class="sig-name descname"><span class="pre">--save_all</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;save_all&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-save_all" title="Link to this definition"></a></dt>
<dd><p>save all variants, not just those passing filters</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-fdr">
<span class="sig-name descname"><span class="pre">--fdr</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;fdr&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-fdr" title="Link to this definition"></a></dt>
<dd><p>FDR for variant calling [0.05]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-min_cov">
<span class="sig-name descname"><span class="pre">--min_cov</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_cov&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-min_cov" title="Link to this definition"></a></dt>
<dd><p>minimum read coverage at potentially variable positions [50]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-gap_window_size">
<span class="sig-name descname"><span class="pre">--gap_window_size</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gap_window_size&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-gap_window_size" title="Link to this definition"></a></dt>
<dd><p>size of window to compute local gap frequency [10]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-max_local_gap_freq">
<span class="sig-name descname"><span class="pre">--max_local_gap_freq</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_local_gap_freq&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-max_local_gap_freq" title="Link to this definition"></a></dt>
<dd><p>max. local gap frequency in window [0.7]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-min_cluster_cov">
<span class="sig-name descname"><span class="pre">--min_cluster_cov</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_cluster_cov&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-min_cluster_cov" title="Link to this definition"></a></dt>
<dd><p>minimum read coverage at potentially variable positions per cluster [10]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-min_freq">
<span class="sig-name descname"><span class="pre">--min_freq</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_freq&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-min_freq" title="Link to this definition"></a></dt>
<dd><p>minimum allele frequency at potentially variable positions (in total or per cluster) [0.40]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-nreads">
<span class="sig-name descname"><span class="pre">--nreads</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nreads&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-nreads" title="Link to this definition"></a></dt>
<dd><p>downsample to this number of reads to control sensitivity</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-variant_annotation">
<span class="sig-name descname"><span class="pre">--variant_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;variant_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-variant_annotation" title="Link to this definition"></a></dt>
<dd><p>variant annotation file (vcf.gz; e.g., from 1000Genomes project)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-haplotypes">
<span class="sig-name descname"><span class="pre">--haplotypes</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;haplotypes&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-haplotypes" title="Link to this definition"></a></dt>
<dd><p>cluster haplotypes</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-find_variants-motifs">
<span class="sig-name descname"><span class="pre">--motifs</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;motifs&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-find_variants-motifs" title="Link to this definition"></a></dt>
<dd><p>comma-separated list of sequence motifs to look up at variant positions [Cg,wrCy,Tw]</p>
</dd></dl>

</section>
<section id="swibrid-get-alignment-pars">
<h3>swibrid get_alignment_pars<a class="headerlink" href="#swibrid-get-alignment-pars" title="Link to this heading"></a></h3>
<p>get LAST-like parameters from a SAM or MAF file. this will go through the alignments and tally mutation and indel frequencies.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_alignment_pars [-h] [-i INF] [-r REF] [-o OUT]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_alignment_pars-h">
<span id="cmdoption-swibrid-get_alignment_pars-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_alignment_pars-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_alignment_pars-i">
<span id="cmdoption-swibrid-get_alignment_pars-inf"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inf&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--inf</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;inf&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_alignment_pars-i" title="Link to this definition"></a></dt>
<dd><p>input .bam or .maf.gz file, or .par from LAST</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_alignment_pars-r">
<span id="cmdoption-swibrid-get_alignment_pars-ref"></span><span class="sig-name descname"><span class="pre">-r</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ref&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ref</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ref&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_alignment_pars-r" title="Link to this definition"></a></dt>
<dd><p>reference genome</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_alignment_pars-o">
<span id="cmdoption-swibrid-get_alignment_pars-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_alignment_pars-o" title="Link to this definition"></a></dt>
<dd><p>output file (.npz)</p>
</dd></dl>

</section>
<section id="swibrid-get-annotation">
<h3>swibrid get_annotation<a class="headerlink" href="#swibrid-get-annotation" title="Link to this heading"></a></h3>
<p>prepare shortened annotation file for insert annotation. input is a GTF file (e.g., from GENCODE), output is a bed file</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_annotation [-h] [-i INPUT] [-o OUT]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_annotation-h">
<span id="cmdoption-swibrid-get_annotation-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_annotation-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_annotation-i">
<span id="cmdoption-swibrid-get_annotation-input"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--input</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;input&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_annotation-i" title="Link to this definition"></a></dt>
<dd><p>input gtf</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_annotation-o">
<span id="cmdoption-swibrid-get_annotation-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_annotation-o" title="Link to this definition"></a></dt>
<dd><p>output bed</p>
</dd></dl>

</section>
<section id="swibrid-get-breakpoint-stats">
<h3>swibrid get_breakpoint_stats<a class="headerlink" href="#swibrid-get-breakpoint-stats" title="Link to this heading"></a></h3>
<p>analyze breakpoint statistics: this will produce aggregate statistics on breakpoints, either averaged over reads, or clusters. homology scores are calculated from binned breakpoint frequencies weighted by homology of bins;
motif scores are calculated from binned breakpoint frequencies weighted by motif occurrences.
output file contains following values:</p>
<ul class="simple">
<li><p>breaks_normalized: number of breaks divided by number of reads / clusters</p></li>
<li><p>frac_breaks_duplications: fraction of breaks leading to duplication events</p></li>
<li><p>frac_breaks_inversions: fraction of breaks leading to inversion events</p></li>
<li><p>frac_breaks_single: fraction of breaks from SM to elsewhere</p></li>
<li><p>frac_breaks_sequential: fraction of breaks to different regions but not SM</p></li>
<li><p>frac_breaks_intra: fraction of breaks within a region</p></li>
<li><p>frac_breaks_inversions/duplications_intra: frac breaks with inversions/duplications within regions</p></li>
<li><p>frac_breaks_X_X: fraction of breaks connecting indicated regions</p></li>
<li><p>spread_XX: spread (standard deviation) of breakpoint positions in a region</p></li>
<li><p>homology_fw: homology in bins around breakpoint positions (same orientation)</p></li>
<li><p>homology_rv: homology in bins around breakpoint positions (opposite orientation)</p></li>
<li><p>homology_fw/rv_XX: homologies for breaks connecting indicated regions</p></li>
<li><p>donor/acceptor_score(_XX): motif scores for donor and acceptor breakpoints for different motifs (subdivided by region)</p></li>
<li><p>donor/acceptor_complexity(_XX): sequence complexity for donoer and acceptor breakpoints (subdivided by region)</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_breakpoint_stats [-h] -g GAPS -r REARRANGEMENTS</span>
<span class="go">                                    [-c CLUSTERING] [-a CLUSTERING_ANALYSIS]</span>
<span class="go">                                    -o OUT [-p PLOT] [-b BINSIZE]</span>
<span class="go">                                    [--scale_factor SCALE_FACTOR]</span>
<span class="go">                                    [--max_gap MAX_GAP] [--homology HOMOLOGY]</span>
<span class="go">                                    [--motifs MOTIFS] [--sample SAMPLE]</span>
<span class="go">                                    [--switch_coords SWITCH_COORDS]</span>
<span class="go">                                    [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                                    [--use_clones USE_CLONES]</span>
<span class="go">                                    [--weights WEIGHTS] [--range RANGE]</span>
<span class="go">                                    [--top_donor TOP_DONOR]</span>
<span class="go">                                    [--top_acceptor TOP_ACCEPTOR] [-n NTOP]</span>
<span class="go">                                    [--reference REFERENCE]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-h">
<span id="cmdoption-swibrid-get_breakpoint_stats-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-g">
<span id="cmdoption-swibrid-get_breakpoint_stats-gaps"></span><span class="sig-name descname"><span class="pre">-g</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gaps&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--gaps</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gaps&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-g" title="Link to this definition"></a></dt>
<dd><p>required: file with gap positions (output of get_gaps)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-r">
<span id="cmdoption-swibrid-get_breakpoint_stats-rearrangements"></span><span class="sig-name descname"><span class="pre">-r</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;rearrangements&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--rearrangements</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;rearrangements&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-r" title="Link to this definition"></a></dt>
<dd><p>required: file with inversion/duplication positions (output of find_rearrangements)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-c">
<span id="cmdoption-swibrid-get_breakpoint_stats-clustering"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--clustering</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-c" title="Link to this definition"></a></dt>
<dd><p>file with clustering results</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-a">
<span id="cmdoption-swibrid-get_breakpoint_stats-clustering_analysis"></span><span class="sig-name descname"><span class="pre">-a</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering_analysis&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--clustering_analysis</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering_analysis&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-a" title="Link to this definition"></a></dt>
<dd><p>file with clustering analysis</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-o">
<span id="cmdoption-swibrid-get_breakpoint_stats-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-o" title="Link to this definition"></a></dt>
<dd><p>required: output file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-p">
<span id="cmdoption-swibrid-get_breakpoint_stats-plot"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;plot&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--plot</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;plot&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-p" title="Link to this definition"></a></dt>
<dd><p>plot file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-b">
<span id="cmdoption-swibrid-get_breakpoint_stats-binsize"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--binsize</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-b" title="Link to this definition"></a></dt>
<dd><p>binsize [50]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-scale_factor">
<span class="sig-name descname"><span class="pre">--scale_factor</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;scale_factor&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-scale_factor" title="Link to this definition"></a></dt>
<dd><p>factor to increase binsize for 2D histogram [10]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-max_gap">
<span class="sig-name descname"><span class="pre">--max_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-max_gap" title="Link to this definition"></a></dt>
<dd><p>max gap size to ignore [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-homology">
<span class="sig-name descname"><span class="pre">--homology</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;homology&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-homology" title="Link to this definition"></a></dt>
<dd><p>file with homology values (output of get_switch_homology)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-motifs">
<span class="sig-name descname"><span class="pre">--motifs</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;motifs&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-motifs" title="Link to this definition"></a></dt>
<dd><p>file with motif counts (output of get_switch_motifs)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-sample">
<span class="sig-name descname"><span class="pre">--sample</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-sample" title="Link to this definition"></a></dt>
<dd><p>sample name (for figure title)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-use_clones">
<span class="sig-name descname"><span class="pre">--use_clones</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;use_clones&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-use_clones" title="Link to this definition"></a></dt>
<dd><p>comma-separated list of clones to use or ‘all’ (default: filtered clusters)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-weights">
<span class="sig-name descname"><span class="pre">--weights</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;weights&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-weights" title="Link to this definition"></a></dt>
<dd><p>use different weights (“cluster” | “reads” | “adjusted”) [cluster]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-range">
<span class="sig-name descname"><span class="pre">--range</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;range&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-range" title="Link to this definition"></a></dt>
<dd><p>range of kmer sizes, e.g., 3,5-7 [5]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-top_donor">
<span class="sig-name descname"><span class="pre">--top_donor</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;top_donor&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-top_donor" title="Link to this definition"></a></dt>
<dd><p>output additional file with top donor sequences</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-top_acceptor">
<span class="sig-name descname"><span class="pre">--top_acceptor</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;top_acceptor&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-top_acceptor" title="Link to this definition"></a></dt>
<dd><p>output additional file with top acceptor sequences</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-n">
<span id="cmdoption-swibrid-get_breakpoint_stats-ntop"></span><span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ntop&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--ntop</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;ntop&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-n" title="Link to this definition"></a></dt>
<dd><p>number of top bins to select</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_breakpoint_stats-reference">
<span class="sig-name descname"><span class="pre">--reference</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;reference&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_breakpoint_stats-reference" title="Link to this definition"></a></dt>
<dd><p>genome fasta file</p>
</dd></dl>

</section>
<section id="swibrid-get-gaps">
<h3>swibrid get_gaps<a class="headerlink" href="#swibrid-get-gaps" title="Link to this heading"></a></h3>
<p>find gaps in (pseudo)MSA: takes the MSA matrix from construct_MSA and finds all gap positions.
output is a .npz file containing several integer arrays</p>
<ul class="simple">
<li><p>read_idx: index of read</p></li>
<li><p>gap_left: left end (within the MSA; 0-based)</p></li>
<li><p>gap_right: right end</p></li>
<li><p>gap_size: size of gap</p></li>
</ul>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_gaps [-h] --msa MSA -o OUT [--paired_end_mode]</span>
<span class="go">                        [--msa_csv MSA_CSV]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_gaps-h">
<span id="cmdoption-swibrid-get_gaps-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_gaps-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_gaps-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_gaps-msa" title="Link to this definition"></a></dt>
<dd><p>required: .npz file with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_gaps-o">
<span id="cmdoption-swibrid-get_gaps-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_gaps-o" title="Link to this definition"></a></dt>
<dd><p>required: output file (.npz)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_gaps-paired_end_mode">
<span class="sig-name descname"><span class="pre">--paired_end_mode</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_gaps-paired_end_mode" title="Link to this definition"></a></dt>
<dd><p>use paired-end mode</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_gaps-msa_csv">
<span class="sig-name descname"><span class="pre">--msa_csv</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa_csv&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_gaps-msa_csv" title="Link to this definition"></a></dt>
<dd><p>msa read info (for paired_end_mode)</p>
</dd></dl>

</section>
<section id="swibrid-get-summary">
<h3>swibrid get_summary<a class="headerlink" href="#swibrid-get-summary" title="Link to this heading"></a></h3>
<p>produce a summary of features derived from a sample and a plot.
collects statistics produced by process_alignments, find_clusters, and downsample_clustering.
produces cluster distribution statistics similar to downsample_clustering (the latter are averaged):</p>
<ul class="simple">
<li><p>mean_cluster_size (as fraction of reads)</p></li>
<li><p>std_cluster_size</p></li>
<li><p>nclusters_final (number of clusters after filtering)</p></li>
<li><p>nclusters_eff (from the entropy of the cluster size distribution pre-filtering)</p></li>
<li><p>cluster_gini: gini coefficient (post-filtering)</p></li>
<li><p>cluster_entropy: entropy (post-filtering)</p></li>
<li><p>cluster_inverse_simpson: inverse simpson coefficient (post-filtering)</p></li>
<li><p>top_clone_occupancy: relative fraction of reads in biggest cluster</p></li>
<li><p>big_clones_occupancy: fraction of reads in clusters &gt; 1% occupancy</p></li>
<li><p>size_length_bias: regression coefficient of log(cluster_size) ~ length</p></li>
<li><p>size_GC_bias: regression coefficient of log(cluster_size) ~ GC</p></li>
</ul>
<p>averages cluster-specific features from analyze_clustering and get_breakpoint_stats over clusters.
collects number of reads / clusters per isotype.
gets statistics on variants (germline vs. somatic, transitions vs. transversions, etc.)</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_summary [-h] [--sample SAMPLE] [--figure FIGURE]</span>
<span class="go">                           [--stats STATS] [--process PROCESS] [--info INFO]</span>
<span class="go">                           [--gaps GAPS] [--breakpoint_stats BREAKPOINT_STATS]</span>
<span class="go">                           [--max_gap MAX_GAP] [--clustering CLUSTERING]</span>
<span class="go">                           [--scanning SCANNING]</span>
<span class="go">                           [--cluster_stats CLUSTER_STATS]</span>
<span class="go">                           [--cluster_analysis CLUSTER_ANALYSIS]</span>
<span class="go">                           [--cluster_downsampling CLUSTER_DOWNSAMPLING]</span>
<span class="go">                           [--variants [VARIANTS]]</span>
<span class="go">                           [--switch_coords SWITCH_COORDS]</span>
<span class="go">                           [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                           [--use_clones USE_CLONES] [--weights WEIGHTS]</span>
<span class="go">                           [--max_n_blunt MAX_N_BLUNT]</span>
<span class="go">                           [--big_clone_cutoff BIG_CLONE_CUTOFF]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-h">
<span id="cmdoption-swibrid-get_summary-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-sample">
<span class="sig-name descname"><span class="pre">--sample</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-sample" title="Link to this definition"></a></dt>
<dd><p>sample name</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-figure">
<span class="sig-name descname"><span class="pre">--figure</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-figure" title="Link to this definition"></a></dt>
<dd><p>output figure</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-stats">
<span class="sig-name descname"><span class="pre">--stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-stats" title="Link to this definition"></a></dt>
<dd><p>output stats</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-process">
<span class="sig-name descname"><span class="pre">--process</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;process&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-process" title="Link to this definition"></a></dt>
<dd><p>stats file from process_alignments</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-info">
<span class="sig-name descname"><span class="pre">--info</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-info" title="Link to this definition"></a></dt>
<dd><p>read info</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-gaps">
<span class="sig-name descname"><span class="pre">--gaps</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;gaps&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-gaps" title="Link to this definition"></a></dt>
<dd><p>gap distribution</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-breakpoint_stats">
<span class="sig-name descname"><span class="pre">--breakpoint_stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;breakpoint_stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-breakpoint_stats" title="Link to this definition"></a></dt>
<dd><p>breakpoint statistics</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-max_gap">
<span class="sig-name descname"><span class="pre">--max_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-max_gap" title="Link to this definition"></a></dt>
<dd><p>max gap size to ignore [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-clustering">
<span class="sig-name descname"><span class="pre">--clustering</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-clustering" title="Link to this definition"></a></dt>
<dd><p>find_clusters output file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-scanning">
<span class="sig-name descname"><span class="pre">--scanning</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;scanning&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-scanning" title="Link to this definition"></a></dt>
<dd><p>find_clusters dendrogram scanning</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-cluster_stats">
<span class="sig-name descname"><span class="pre">--cluster_stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cluster_stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-cluster_stats" title="Link to this definition"></a></dt>
<dd><p>find_clusters stats</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-cluster_analysis">
<span class="sig-name descname"><span class="pre">--cluster_analysis</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cluster_analysis&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-cluster_analysis" title="Link to this definition"></a></dt>
<dd><p>analyze_clusters output</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-cluster_downsampling">
<span class="sig-name descname"><span class="pre">--cluster_downsampling</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cluster_downsampling&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-cluster_downsampling" title="Link to this definition"></a></dt>
<dd><p>downsample_clusters output</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-variants">
<span class="sig-name descname"><span class="pre">--variants</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;variants&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-variants" title="Link to this definition"></a></dt>
<dd><p>file with variant table (from find_variants)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-use_clones">
<span class="sig-name descname"><span class="pre">--use_clones</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;use_clones&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-use_clones" title="Link to this definition"></a></dt>
<dd><p>comma-separated list of clones to use</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-weights">
<span class="sig-name descname"><span class="pre">--weights</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;weights&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-weights" title="Link to this definition"></a></dt>
<dd><p>specify weights (“cluster” | “reads” | “adjusted”) [cluster]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-max_n_blunt">
<span class="sig-name descname"><span class="pre">--max_n_blunt</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_n_blunt&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-max_n_blunt" title="Link to this definition"></a></dt>
<dd><p>max avg. number of untemplated / homologous nucleotides of reads in cluster to be considered a blunt end [0.5]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_summary-big_clone_cutoff">
<span class="sig-name descname"><span class="pre">--big_clone_cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;big_clone_cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_summary-big_clone_cutoff" title="Link to this definition"></a></dt>
<dd><p>cutoff to determine a “big” clone as fraction of clustered reads [0.01]</p>
</dd></dl>

</section>
<section id="swibrid-get-switch-homology">
<h3>swibrid get_switch_homology<a class="headerlink" href="#swibrid-get-switch-homology" title="Link to this heading"></a></h3>
<p>analyze sequence homology within switch region bins by jaccard similarity of k-mers:
from switch region coordinates and a genome file, this script assesses homology of
pairs of bins by jaccard similarity of k-mer occurrences in forward or reverse orientation.
output is a .npz file with pairwise forward and reverse homology arrays for each value of k</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_switch_homology [-h] -g GENOME -o OUTPUT</span>
<span class="go">                                   --switch_annotation SWITCH_ANNOTATION</span>
<span class="go">                                   [--switch_coords SWITCH_COORDS]</span>
<span class="go">                                   [-b BINSIZE] [--range RANGE]</span>
<span class="go">                                   [--figure FIGURE]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-h">
<span id="cmdoption-swibrid-get_switch_homology-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-g">
<span id="cmdoption-swibrid-get_switch_homology-genome"></span><span class="sig-name descname"><span class="pre">-g</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;genome&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--genome</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;genome&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-g" title="Link to this definition"></a></dt>
<dd><p>required: genome fasta file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-o">
<span id="cmdoption-swibrid-get_switch_homology-output"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-o" title="Link to this definition"></a></dt>
<dd><p>required: output file (npz)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>required: bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-b">
<span id="cmdoption-swibrid-get_switch_homology-binsize"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--binsize</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-b" title="Link to this definition"></a></dt>
<dd><p>binsize [100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-range">
<span class="sig-name descname"><span class="pre">--range</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;range&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-range" title="Link to this definition"></a></dt>
<dd><p>range of kmer sizes, e.g., 3,5-7 [5]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_homology-figure">
<span class="sig-name descname"><span class="pre">--figure</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_homology-figure" title="Link to this definition"></a></dt>
<dd><p>output figure</p>
</dd></dl>

</section>
<section id="swibrid-get-switch-motifs">
<h3>swibrid get_switch_motifs<a class="headerlink" href="#swibrid-get-switch-motifs" title="Link to this heading"></a></h3>
<p>analyze sequence content within switch region bins by a number of motifs.
given a list of motifs, genome file and switch region coordinates, this script
counts the occurrences of these motifs in coordinate bins.
output is a .npz file with count arrays for each motif</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_switch_motifs [-h] -g GENOME -o OUTPUT</span>
<span class="go">                                 [--switch_coords SWITCH_COORDS]</span>
<span class="go">                                 [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                                 [-b BINSIZE] [--motifs MOTIFS] [-k K]</span>
<span class="go">                                 [--figure FIGURE]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-h">
<span id="cmdoption-swibrid-get_switch_motifs-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-g">
<span id="cmdoption-swibrid-get_switch_motifs-genome"></span><span class="sig-name descname"><span class="pre">-g</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;genome&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--genome</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;genome&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-g" title="Link to this definition"></a></dt>
<dd><p>required: genome fasta file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-o">
<span id="cmdoption-swibrid-get_switch_motifs-output"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-o" title="Link to this definition"></a></dt>
<dd><p>required: output file (npz)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000:-]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-b">
<span id="cmdoption-swibrid-get_switch_motifs-binsize"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--binsize</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;binsize&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-b" title="Link to this definition"></a></dt>
<dd><p>binsize [100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-motifs">
<span class="sig-name descname"><span class="pre">--motifs</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;motifs&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-motifs" title="Link to this definition"></a></dt>
<dd><p>comma-separated list of sequence motifs [S=G/C,W=A/T,WGCW,GAGCT,GGGST]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-k">
<span class="sig-name descname"><span class="pre">-k</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;k&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-k" title="Link to this definition"></a></dt>
<dd><p>value k for estimate of k-mer complexity [5]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_switch_motifs-figure">
<span class="sig-name descname"><span class="pre">--figure</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_switch_motifs-figure" title="Link to this definition"></a></dt>
<dd><p>output figure</p>
</dd></dl>

</section>
<section id="swibrid-get-synthetic-reads">
<h3>swibrid get_synthetic_reads<a class="headerlink" href="#swibrid-get-synthetic-reads" title="Link to this heading"></a></h3>
<p>create synthetic reads from bed file.
mutations, insertions and deletions will be added according to parameters estimated by LAST.
clone sizes can be uniform or distributed as Poisson or Negative Binomial.
additional homozygous / heterozygous / other variants can be added</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_synthetic_reads [-h] -b BED -r REFERENCE -p PAR -o OUT</span>
<span class="go">                                   [-i INFO] [-n N] [-k K] [-d DISTRIBUTION]</span>
<span class="go">                                   [--nbinom_alpha NBINOM_ALPHA]</span>
<span class="go">                                   [--variants [VARIANTS]]</span>
<span class="go">                                   [--rearrangements [REARRANGEMENTS]]</span>
<span class="go">                                   [-s SEED] [--no_mutations] [--no_deletions]</span>
<span class="go">                                   [--no_insertions]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-h">
<span id="cmdoption-swibrid-get_synthetic_reads-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-b">
<span id="cmdoption-swibrid-get_synthetic_reads-bed"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--bed</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-b" title="Link to this definition"></a></dt>
<dd><p>required: input bed file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-r">
<span id="cmdoption-swibrid-get_synthetic_reads-reference"></span><span class="sig-name descname"><span class="pre">-r</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;reference&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--reference</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;reference&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-r" title="Link to this definition"></a></dt>
<dd><p>required: reference sequence</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-p">
<span id="cmdoption-swibrid-get_synthetic_reads-par"></span><span class="sig-name descname"><span class="pre">-p</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;par&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--par</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;par&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-p" title="Link to this definition"></a></dt>
<dd><p>required: file with LAST parameters</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-o">
<span id="cmdoption-swibrid-get_synthetic_reads-out"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--out</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;out&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-o" title="Link to this definition"></a></dt>
<dd><p>required: output fasta / fastq (.gz) file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-i">
<span id="cmdoption-swibrid-get_synthetic_reads-info"></span><span class="sig-name descname"><span class="pre">-i</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--info</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-i" title="Link to this definition"></a></dt>
<dd><p>output info file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-n">
<span id="cmdoption-swibrid-get_synthetic_reads-n"></span><span class="sig-name descname"><span class="pre">-n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--n</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;n&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-n" title="Link to this definition"></a></dt>
<dd><p>use n random entries of input bed file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-k">
<span id="cmdoption-swibrid-get_synthetic_reads-k"></span><span class="sig-name descname"><span class="pre">-k</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;k&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--k</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;k&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-k" title="Link to this definition"></a></dt>
<dd><p>number of mutated copies per sequence [1.0]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-d">
<span id="cmdoption-swibrid-get_synthetic_reads-distribution"></span><span class="sig-name descname"><span class="pre">-d</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;distribution&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--distribution</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;distribution&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-d" title="Link to this definition"></a></dt>
<dd><p>distribution type for k (delta | poisson | nbinom | P | NB | D) [poisson]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-nbinom_alpha">
<span class="sig-name descname"><span class="pre">--nbinom_alpha</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;nbinom_alpha&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-nbinom_alpha" title="Link to this definition"></a></dt>
<dd><p>dispersion alpha of negative binomial [4.0]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-variants">
<span class="sig-name descname"><span class="pre">--variants</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;variants&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-variants" title="Link to this definition"></a></dt>
<dd><p>variant file (like output of find_variants) to add variants</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-rearrangements">
<span class="sig-name descname"><span class="pre">--rearrangements</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;rearrangements&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-rearrangements" title="Link to this definition"></a></dt>
<dd><p>rearrangement file (bed-like output of find_rearrangements) to add rearrangements</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-s">
<span id="cmdoption-swibrid-get_synthetic_reads-seed"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;seed&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--seed</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;seed&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-s" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-no_mutations">
<span class="sig-name descname"><span class="pre">--no_mutations</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-no_mutations" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-no_deletions">
<span class="sig-name descname"><span class="pre">--no_deletions</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-no_deletions" title="Link to this definition"></a></dt>
<dd></dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_synthetic_reads-no_insertions">
<span class="sig-name descname"><span class="pre">--no_insertions</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_synthetic_reads-no_insertions" title="Link to this definition"></a></dt>
<dd></dd></dl>

</section>
<section id="swibrid-get-unique-clones-bed">
<h3>swibrid get_unique_clones_bed<a class="headerlink" href="#swibrid-get-unique-clones-bed" title="Link to this heading"></a></h3>
<p>produce a bed file with unique clones from the biggest clusters.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid get_unique_clones_bed [-h] -b BED -c CLUSTERING [-o OUTPUT]</span>
<span class="go">                                     [--cut CUT]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_unique_clones_bed-h">
<span id="cmdoption-swibrid-get_unique_clones_bed-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-get_unique_clones_bed-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_unique_clones_bed-b">
<span id="cmdoption-swibrid-get_unique_clones_bed-bed"></span><span class="sig-name descname"><span class="pre">-b</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--bed</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;bed&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_unique_clones_bed-b" title="Link to this definition"></a></dt>
<dd><p>required: input bed file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_unique_clones_bed-c">
<span id="cmdoption-swibrid-get_unique_clones_bed-clustering"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--clustering</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_unique_clones_bed-c" title="Link to this definition"></a></dt>
<dd><p>required: clustering results</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_unique_clones_bed-o">
<span id="cmdoption-swibrid-get_unique_clones_bed-output"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--output</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;output&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_unique_clones_bed-o" title="Link to this definition"></a></dt>
<dd><p>output file [stdout]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-get_unique_clones_bed-cut">
<span class="sig-name descname"><span class="pre">--cut</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cut&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-get_unique_clones_bed-cut" title="Link to this definition"></a></dt>
<dd><p>use only this fraction of the biggest clusters [0.8]</p>
</dd></dl>

</section>
<section id="swibrid-plot-clustering">
<h3>swibrid plot_clustering<a class="headerlink" href="#swibrid-plot-clustering" title="Link to this heading"></a></h3>
<p>Given a MSA and clustering results, this will display individual reads mapping over the switch regions.
reads will be ordered as dictated by the linkage, or simply by isotype and cluster value.
reads can be colored by different variables (isotype, cluster, haplotype, coverage, strand, orientation or other columns present in the info file).
an additional sidebar can display additional variables.
variant positions or breakpoint realignment statistics can also be indicated.</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid plot_clustering [-h] --msa MSA --figure FIGURE</span>
<span class="go">                               --clustering_results CLUSTERING_RESULTS</span>
<span class="go">                               --switch_annotation SWITCH_ANNOTATION</span>
<span class="go">                               [--linkage LINKAGE] [--sample SAMPLE]</span>
<span class="go">                               [--info INFO]</span>
<span class="go">                               [--clustering_stats CLUSTERING_STATS]</span>
<span class="go">                               [--cutoff CUTOFF]</span>
<span class="go">                               [--switch_coords SWITCH_COORDS]</span>
<span class="go">                               [--annotation [ANNOTATION]]</span>
<span class="go">                               [--color_by COLOR_BY]</span>
<span class="go">                               [--sidebar_color_by SIDEBAR_COLOR_BY]</span>
<span class="go">                               [--show_inserts] [--coords COORDS]</span>
<span class="go">                               [--no_x_ticks] [--omit_scale_bar]</span>
<span class="go">                               [--fig_width FIG_WIDTH]</span>
<span class="go">                               [--fig_height FIG_HEIGHT]</span>
<span class="go">                               [--linkage_border LINKAGE_BORDER]</span>
<span class="go">                               [--cutoff_color CUTOFF_COLOR]</span>
<span class="go">                               [--variants_table VARIANTS_TABLE]</span>
<span class="go">                               [--variants_matrix VARIANTS_MATRIX]</span>
<span class="go">                               [--haplotypes HAPLOTYPES]</span>
<span class="go">                               [--realignments REALIGNMENTS] [--dpi DPI]</span>
<span class="go">                               [--chunksize CHUNKSIZE] [--cmax CMAX]</span>
<span class="go">                               [--paired_end_mode]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-h">
<span id="cmdoption-swibrid-plot_clustering-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-msa">
<span class="sig-name descname"><span class="pre">--msa</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;msa&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-msa" title="Link to this definition"></a></dt>
<dd><p>required: file(s) with  (pseudo) multiple alignment of read sequences for clustering</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-figure">
<span class="sig-name descname"><span class="pre">--figure</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-figure" title="Link to this definition"></a></dt>
<dd><p>required: output figure</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-clustering_results">
<span class="sig-name descname"><span class="pre">--clustering_results</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering_results&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-clustering_results" title="Link to this definition"></a></dt>
<dd><p>required: file contains clustering results</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>BED file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-linkage">
<span class="sig-name descname"><span class="pre">--linkage</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;linkage&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-linkage" title="Link to this definition"></a></dt>
<dd><p>file containing linkage</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-sample">
<span class="sig-name descname"><span class="pre">--sample</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-sample" title="Link to this definition"></a></dt>
<dd><p>sample name (for figure title)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-info">
<span class="sig-name descname"><span class="pre">--info</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-info" title="Link to this definition"></a></dt>
<dd><p>file with read info</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-clustering_stats">
<span class="sig-name descname"><span class="pre">--clustering_stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;clustering_stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-clustering_stats" title="Link to this definition"></a></dt>
<dd><p>file contains clustering stats with cutoff value</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-cutoff">
<span class="sig-name descname"><span class="pre">--cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-cutoff" title="Link to this definition"></a></dt>
<dd><p>show cutoff line</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-switch_coords" title="Link to this definition"></a></dt>
<dd><p>Coordinates of switch region [chr14:106050000-106337000:-].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-annotation">
<span class="sig-name descname"><span class="pre">--annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with gene annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-color_by">
<span class="sig-name descname"><span class="pre">--color_by</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;color_by&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-color_by" title="Link to this definition"></a></dt>
<dd><p>Color reads by isotype, cluster, haplotype, or other columns in the ‘info’ file [isotype].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-sidebar_color_by">
<span class="sig-name descname"><span class="pre">--sidebar_color_by</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sidebar_color_by&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-sidebar_color_by" title="Link to this definition"></a></dt>
<dd><p>Color sidebar reads by (comma-separated list of) isotype, cluster, haplotype, or other columns in the ‘info’ file [none].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-show_inserts">
<span class="sig-name descname"><span class="pre">--show_inserts</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-show_inserts" title="Link to this definition"></a></dt>
<dd><p>Show insert locations.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-coords">
<span class="sig-name descname"><span class="pre">--coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-coords" title="Link to this definition"></a></dt>
<dd><p>file with processed read coordinates (required if inserts are shown)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-no_x_ticks">
<span class="sig-name descname"><span class="pre">--no_x_ticks</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-no_x_ticks" title="Link to this definition"></a></dt>
<dd><p>Omit the x-axis ticks.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-omit_scale_bar">
<span class="sig-name descname"><span class="pre">--omit_scale_bar</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-omit_scale_bar" title="Link to this definition"></a></dt>
<dd><p>Omit scale bar.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-fig_width">
<span class="sig-name descname"><span class="pre">--fig_width</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;fig_width&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-fig_width" title="Link to this definition"></a></dt>
<dd><p>Height of figure in inches [6].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-fig_height">
<span class="sig-name descname"><span class="pre">--fig_height</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;fig_height&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-fig_height" title="Link to this definition"></a></dt>
<dd><p>width of figure in inches [8]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-linkage_border">
<span class="sig-name descname"><span class="pre">--linkage_border</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;linkage_border&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-linkage_border" title="Link to this definition"></a></dt>
<dd><p>Fraction of figure used for dendrogram [0.10].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-cutoff_color">
<span class="sig-name descname"><span class="pre">--cutoff_color</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cutoff_color&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-cutoff_color" title="Link to this definition"></a></dt>
<dd><p>Color of dendrogram cutoff line [r].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-variants_table">
<span class="sig-name descname"><span class="pre">--variants_table</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;variants_table&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-variants_table" title="Link to this definition"></a></dt>
<dd><p>Show variant positions from variant table (from find_variants).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-variants_matrix">
<span class="sig-name descname"><span class="pre">--variants_matrix</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;variants_matrix&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-variants_matrix" title="Link to this definition"></a></dt>
<dd><p>Indicate variant occurrences from variant matrix (from find_variants).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-haplotypes">
<span class="sig-name descname"><span class="pre">--haplotypes</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;haplotypes&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-haplotypes" title="Link to this definition"></a></dt>
<dd><p>File with haplotype clustering (from find_variants).</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-realignments">
<span class="sig-name descname"><span class="pre">--realignments</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;realignments&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-realignments" title="Link to this definition"></a></dt>
<dd><p>Indicate realignment scores with breakpoint realignments.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-dpi">
<span class="sig-name descname"><span class="pre">--dpi</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;dpi&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-dpi" title="Link to this definition"></a></dt>
<dd><p>DPI of output figure [300].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-chunksize">
<span class="sig-name descname"><span class="pre">--chunksize</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;chunksize&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-chunksize" title="Link to this definition"></a></dt>
<dd><p>Plot image in chunks of reads of this size [5000].</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-cmax">
<span class="sig-name descname"><span class="pre">--cmax</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;cmax&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-cmax" title="Link to this definition"></a></dt>
<dd><p>Maximum height in dendrogram plot.</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_clustering-paired_end_mode">
<span class="sig-name descname"><span class="pre">--paired_end_mode</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-plot_clustering-paired_end_mode" title="Link to this definition"></a></dt>
<dd><p>Use paired-end mode (requires coords file to indicate links).</p>
</dd></dl>

</section>
<section id="swibrid-plot-demux-report">
<h3>swibrid plot_demux_report<a class="headerlink" href="#swibrid-plot-demux-report" title="Link to this heading"></a></h3>
<p>plot demultiplexing report</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid plot_demux_report [-h] [-o OUTDIR] [-f FIGURE]</span>
<span class="go">                                 [-s SAMPLE_SHEET]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_demux_report-h">
<span id="cmdoption-swibrid-plot_demux_report-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-plot_demux_report-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_demux_report-o">
<span id="cmdoption-swibrid-plot_demux_report-outdir"></span><span class="sig-name descname"><span class="pre">-o</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;outdir&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--outdir</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;outdir&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_demux_report-o" title="Link to this definition"></a></dt>
<dd><p>output directory</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_demux_report-f">
<span id="cmdoption-swibrid-plot_demux_report-figure"></span><span class="sig-name descname"><span class="pre">-f</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--figure</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;figure&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_demux_report-f" title="Link to this definition"></a></dt>
<dd><p>summary figure</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-plot_demux_report-s">
<span id="cmdoption-swibrid-plot_demux_report-sample_sheet"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample_sheet&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--sample_sheet</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sample_sheet&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-plot_demux_report-s" title="Link to this definition"></a></dt>
<dd><p>sample sheet (barcode &lt;tab&gt; sample_name, no header)</p>
</dd></dl>

</section>
<section id="swibrid-process-alignments">
<h3>swibrid process_alignments<a class="headerlink" href="#swibrid-process-alignments" title="Link to this heading"></a></h3>
<p>process alignments: for an input file with aligned reads (MAF if LAST output, SAM if minimap2 output),
get alignments to switch regions and elsewhere (potential inserts).</p>
<p>reads are removed if</p>
<ul class="simple">
<li><p>they are too short</p></li>
<li><p>they don’t contain a forward and reverse primer</p></li>
<li><p>they contain internal primers</p></li>
<li><p>they contain no switch region</p></li>
<li><p>if mapped regions on the genome are much shorter than mapped parts of the read</p></li>
<li><p>there’s too much overlap between different alignments on the read, or on the genome</p></li>
<li><p>too little of the read maps</p></li>
<li><p>alignments to the switch region are in the wrong order</p></li>
<li><p>the isotype cannot be determined</p></li>
</ul>
<p>output table contains</p>
<ul class="simple">
<li><p>isotype</p></li>
<li><p>read orientation</p></li>
<li><p>read coverage</p></li>
<li><p>fraction of read sequence mapping to the same genomic regions</p></li>
<li><p>mapped switch region segments</p></li>
<li><p>inserts</p></li>
</ul>
<p>aligned sequences can be saved separately (necessary to then construct a pseudo MSA).
if <cite>–realign_breakpoints</cite> is set, 20nt on each side of a breakpoint are re-aligned, and statistics like number of homologous or untemplated bases are extracted</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid process_alignments [-h] --alignments ALIGNMENTS --outfile</span>
<span class="go">                                  OUTFILE --info INFO [--sequences SEQUENCES]</span>
<span class="go">                                  [--stats STATS]</span>
<span class="go">                                  [--switch_coords SWITCH_COORDS]</span>
<span class="go">                                  [--switch_annotation SWITCH_ANNOTATION]</span>
<span class="go">                                  [--min_length MIN_LENGTH] [--only_complete]</span>
<span class="go">                                  [--keep_internal] [--remove_duplicates]</span>
<span class="go">                                  [--ignore_order] [--telo TELO]</span>
<span class="go">                                  [--min_switch_match MIN_SWITCH_MATCH]</span>
<span class="go">                                  [--max_switch_overlap MAX_SWITCH_OVERLAP]</span>
<span class="go">                                  [--max_insert_gap MAX_INSERT_GAP]</span>
<span class="go">                                  [--max_insert_overlap MAX_INSERT_OVERLAP]</span>
<span class="go">                                  [--min_insert_match MIN_INSERT_MATCH]</span>
<span class="go">                                  [--min_num_switch_matches MIN_NUM_SWITCH_MATCHES]</span>
<span class="go">                                  [--min_cov MIN_COV] [--max_gap MAX_GAP]</span>
<span class="go">                                  [--isotype_extra ISOTYPE_EXTRA]</span>
<span class="go">                                  [--telo_cutoff TELO_CUTOFF]</span>
<span class="go">                                  [--min_telo_matchlen MIN_TELO_MATCHLEN]</span>
<span class="go">                                  [--blacklist_regions [BLACKLIST_REGIONS]]</span>
<span class="go">                                  [--realign_breakpoints REALIGN_BREAKPOINTS]</span>
<span class="go">                                  [--realignment_penalties REALIGNMENT_PENALTIES]</span>
<span class="go">                                  [--raw_reads RAW_READS] [--genome GENOME]</span>
<span class="go">                                  [--keep_secondary_alignments]</span>
<span class="go">                                  [--paired_end_mode]</span>
<span class="go">                                  [--remove_isotypes REMOVE_ISOTYPES]</span>
<span class="go">                                  [--interrupt_for_read INTERRUPT_FOR_READ]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-h">
<span id="cmdoption-swibrid-process_alignments-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-alignments">
<span class="sig-name descname"><span class="pre">--alignments</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;alignments&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-alignments" title="Link to this definition"></a></dt>
<dd><p>required: MAF file (LAST output) or SAM file (minimap2 output) with aligned reads (LAST output)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-outfile">
<span class="sig-name descname"><span class="pre">--outfile</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;outfile&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-outfile" title="Link to this definition"></a></dt>
<dd><p>required: output table with processed reads</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-info">
<span class="sig-name descname"><span class="pre">--info</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;info&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-info" title="Link to this definition"></a></dt>
<dd><p>required: csv file with read info</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-sequences">
<span class="sig-name descname"><span class="pre">--sequences</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;sequences&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-sequences" title="Link to this definition"></a></dt>
<dd><p>save sequence alignment to reference for pseudo-multiple alignment</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-stats">
<span class="sig-name descname"><span class="pre">--stats</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;stats&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-stats" title="Link to this definition"></a></dt>
<dd><p>output stats on removed reads</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-switch_coords">
<span class="sig-name descname"><span class="pre">--switch_coords</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_coords&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-switch_coords" title="Link to this definition"></a></dt>
<dd><p>coordinates of switch region [chr14:106050000-106337000]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-switch_annotation">
<span class="sig-name descname"><span class="pre">--switch_annotation</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;switch_annotation&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-switch_annotation" title="Link to this definition"></a></dt>
<dd><p>bed file with switch annotation</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-min_length">
<span class="sig-name descname"><span class="pre">--min_length</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_length&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-min_length" title="Link to this definition"></a></dt>
<dd><p>minimum read length [500]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-only_complete">
<span class="sig-name descname"><span class="pre">--only_complete</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-only_complete" title="Link to this definition"></a></dt>
<dd><p>use only complete reads with two primers</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-keep_internal">
<span class="sig-name descname"><span class="pre">--keep_internal</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-keep_internal" title="Link to this definition"></a></dt>
<dd><p>keep reads with internal primers</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-remove_duplicates">
<span class="sig-name descname"><span class="pre">--remove_duplicates</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-remove_duplicates" title="Link to this definition"></a></dt>
<dd><p>remove duplicate reads (based on UMIs in read IDs)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-ignore_order">
<span class="sig-name descname"><span class="pre">--ignore_order</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-ignore_order" title="Link to this definition"></a></dt>
<dd><p>don’t check order of matches along read</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-telo">
<span class="sig-name descname"><span class="pre">--telo</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;telo&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-telo" title="Link to this definition"></a></dt>
<dd><p>output of blasting reads against telomer repeats</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-min_switch_match">
<span class="sig-name descname"><span class="pre">--min_switch_match</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_switch_match&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-min_switch_match" title="Link to this definition"></a></dt>
<dd><p>min match length to switch region</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-max_switch_overlap">
<span class="sig-name descname"><span class="pre">--max_switch_overlap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_switch_overlap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-max_switch_overlap" title="Link to this definition"></a></dt>
<dd><p>max overlap between switch parts of different orientation [0]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-max_insert_gap">
<span class="sig-name descname"><span class="pre">--max_insert_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_insert_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-max_insert_gap" title="Link to this definition"></a></dt>
<dd><p>max gap between insert and switch parts (on the read) [100]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-max_insert_overlap">
<span class="sig-name descname"><span class="pre">--max_insert_overlap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_insert_overlap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-max_insert_overlap" title="Link to this definition"></a></dt>
<dd><p>max overlap between switch and insert parts (on the read) [50]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-min_insert_match">
<span class="sig-name descname"><span class="pre">--min_insert_match</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_insert_match&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-min_insert_match" title="Link to this definition"></a></dt>
<dd><p>min length of insert [50]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-min_num_switch_matches">
<span class="sig-name descname"><span class="pre">--min_num_switch_matches</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_num_switch_matches&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-min_num_switch_matches" title="Link to this definition"></a></dt>
<dd><p>min number of switch matches in read [1]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-min_cov">
<span class="sig-name descname"><span class="pre">--min_cov</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_cov&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-min_cov" title="Link to this definition"></a></dt>
<dd><p>coverage cutoff on reads [0.90]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-max_gap">
<span class="sig-name descname"><span class="pre">--max_gap</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;max_gap&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-max_gap" title="Link to this definition"></a></dt>
<dd><p>split alignments with gaps beyond that value [75]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-isotype_extra">
<span class="sig-name descname"><span class="pre">--isotype_extra</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;isotype_extra&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-isotype_extra" title="Link to this definition"></a></dt>
<dd><p>extra space added to define isotypes [200]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-telo_cutoff">
<span class="sig-name descname"><span class="pre">--telo_cutoff</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;telo_cutoff&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-telo_cutoff" title="Link to this definition"></a></dt>
<dd><p>% identify cutoff for telomer match [90]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-min_telo_matchlen">
<span class="sig-name descname"><span class="pre">--min_telo_matchlen</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;min_telo_matchlen&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-min_telo_matchlen" title="Link to this definition"></a></dt>
<dd><p>minimum matchlen for telomeric repeat matches [25]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-blacklist_regions">
<span class="sig-name descname"><span class="pre">--blacklist_regions</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;blacklist_regions&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-blacklist_regions" title="Link to this definition"></a></dt>
<dd><p>ignore inserts from blacklist regions defined in this bed file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-realign_breakpoints">
<span class="sig-name descname"><span class="pre">--realign_breakpoints</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;realign_breakpoints&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-realign_breakpoints" title="Link to this definition"></a></dt>
<dd><p>re-align reads around breakpoints and save results to this file (requires <code class="docutils literal notranslate"><span class="pre">--raw_reads</span></code> and <code class="docutils literal notranslate"><span class="pre">--genome</span></code>)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-realignment_penalties">
<span class="sig-name descname"><span class="pre">--realignment_penalties</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;realignment_penalties&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-realignment_penalties" title="Link to this definition"></a></dt>
<dd><p>re-alignment penalties (“ont”, “hifi” or “sr”) [ont]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-raw_reads">
<span class="sig-name descname"><span class="pre">--raw_reads</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;raw_reads&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-raw_reads" title="Link to this definition"></a></dt>
<dd><p>fasta file with unaligned reads (interleaved mates for paired-end mode)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-genome">
<span class="sig-name descname"><span class="pre">--genome</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;genome&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-genome" title="Link to this definition"></a></dt>
<dd><p>reference genome file</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-keep_secondary_alignments">
<span class="sig-name descname"><span class="pre">--keep_secondary_alignments</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-keep_secondary_alignments" title="Link to this definition"></a></dt>
<dd><p>keep secondary alignments in SAM file (but conflicting alignments are not resolved!)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-paired_end_mode">
<span class="sig-name descname"><span class="pre">--paired_end_mode</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-paired_end_mode" title="Link to this definition"></a></dt>
<dd><p>EXPERIMENTAL: use paired-end mode (use interleaved mates as fastq input)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-remove_isotypes">
<span class="sig-name descname"><span class="pre">--remove_isotypes</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;remove_isotypes&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-remove_isotypes" title="Link to this definition"></a></dt>
<dd><p>remove reads with these isotypes (comma-separated list)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-process_alignments-interrupt_for_read">
<span class="sig-name descname"><span class="pre">--interrupt_for_read</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;interrupt_for_read&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-process_alignments-interrupt_for_read" title="Link to this definition"></a></dt>
<dd><p>(for debugging) interrupt for specified read(s)</p>
</dd></dl>

</section>
<section id="swibrid-run">
<h3>swibrid run<a class="headerlink" href="#swibrid-run" title="Link to this heading"></a></h3>
<p>main command: run the entire pipeline.
run <code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">setup</span></code> and edit config.yaml first!</p>
<p>this will call snakemake; additional options are passed to snakemake
e.g., for a dry run use <code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">run</span> <span class="pre">-n</span></code>, or <code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">run</span> <span class="pre">--unlock</span></code> after a failed run</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid run [-h] [-s SNAKEFILE] [-c CONFIG] [--slurm] [--sge]</span>
<span class="go">                   [--queue QUEUE]</span>
<span class="go">                   ...</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-arg-snake_options">
<span class="sig-name descname"><span class="pre">snake_options</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-run-arg-snake_options" title="Link to this definition"></a></dt>
<dd><p>pass options to snakemake (…)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-h">
<span id="cmdoption-swibrid-run-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-run-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-s">
<span id="cmdoption-swibrid-run-snakefile"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;snakefile&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--snakefile</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;snakefile&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-run-s" title="Link to this definition"></a></dt>
<dd><p>snakefile [pipeline.snake]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-c">
<span id="cmdoption-swibrid-run-config"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;config&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--config</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;config&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-run-c" title="Link to this definition"></a></dt>
<dd><p>config file [config.yaml]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-slurm">
<span class="sig-name descname"><span class="pre">--slurm</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-run-slurm" title="Link to this definition"></a></dt>
<dd><p>submit to slurm</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-sge">
<span class="sig-name descname"><span class="pre">--sge</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-run-sge" title="Link to this definition"></a></dt>
<dd><p>submit to SGE</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-run-queue">
<span class="sig-name descname"><span class="pre">--queue</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;queue&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-run-queue" title="Link to this definition"></a></dt>
<dd><p>slurm queue [medium]</p>
</dd></dl>

</section>
<section id="swibrid-setup">
<h3>swibrid setup<a class="headerlink" href="#swibrid-setup" title="Link to this heading"></a></h3>
<p>main command: copy config and snakefile to current directory</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid setup [-h] [-f]</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-setup-h">
<span id="cmdoption-swibrid-setup-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-setup-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-setup-f">
<span id="cmdoption-swibrid-setup-overwrite"></span><span class="sig-name descname"><span class="pre">-f</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--overwrite</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-setup-f" title="Link to this definition"></a></dt>
<dd><p>overwrite files if present [no]</p>
</dd></dl>

</section>
<section id="swibrid-test">
<h3>swibrid test<a class="headerlink" href="#swibrid-test" title="Link to this heading"></a></h3>
<p>test the pipeline.
this will create synthetic reads in <code class="docutils literal notranslate"><span class="pre">input</span></code> and run the pipeline on this data,
using a reduced hg38 genome in <code class="docutils literal notranslate"><span class="pre">index</span></code> with only the switch region (chr14:105000000-106000000).
it will probably take about 30-60 minutes and will call snakemake, passing on additional options.
e.g., for a dry run use <code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">test</span> <span class="pre">-n</span></code>, and do <code class="docutils literal notranslate"><span class="pre">swibrid</span> <span class="pre">test</span> <span class="pre">--unlock</span></code> after a failed run</p>
<div class="highlight-console notranslate"><div class="highlight"><pre><span></span><span class="go">usage: swibrid test [-h] [-f] [-s SNAKEFILE] [-c CONFIG] [--slurm] [--sge]</span>
<span class="go">                    [--queue QUEUE]</span>
<span class="go">                    ...</span>
</pre></div>
</div>
<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-arg-snake_options">
<span class="sig-name descname"><span class="pre">snake_options</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-test-arg-snake_options" title="Link to this definition"></a></dt>
<dd><p>pass options to snakemake (…)</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-h">
<span id="cmdoption-swibrid-test-help"></span><span class="sig-name descname"><span class="pre">-h</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--help</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-test-h" title="Link to this definition"></a></dt>
<dd><p>show this help message and exit</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-f">
<span id="cmdoption-swibrid-test-overwrite"></span><span class="sig-name descname"><span class="pre">-f</span></span><span class="sig-prename descclassname"></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--overwrite</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-test-f" title="Link to this definition"></a></dt>
<dd><p>overwrite files if present [no]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-s">
<span id="cmdoption-swibrid-test-snakefile"></span><span class="sig-name descname"><span class="pre">-s</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;snakefile&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--snakefile</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;snakefile&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-test-s" title="Link to this definition"></a></dt>
<dd><p>snakefile [pipeline.snake]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-c">
<span id="cmdoption-swibrid-test-config"></span><span class="sig-name descname"><span class="pre">-c</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;config&gt;</span></span><span class="sig-prename descclassname"><span class="pre">,</span> </span><span class="sig-name descname"><span class="pre">--config</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;config&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-test-c" title="Link to this definition"></a></dt>
<dd><p>config file [test_config.yaml]</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-slurm">
<span class="sig-name descname"><span class="pre">--slurm</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-test-slurm" title="Link to this definition"></a></dt>
<dd><p>submit to slurm</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-sge">
<span class="sig-name descname"><span class="pre">--sge</span></span><span class="sig-prename descclassname"></span><a class="headerlink" href="#cmdoption-swibrid-test-sge" title="Link to this definition"></a></dt>
<dd><p>submit to SGE</p>
</dd></dl>

<dl class="std option">
<dt class="sig sig-object std" id="cmdoption-swibrid-test-queue">
<span class="sig-name descname"><span class="pre">--queue</span></span><span class="sig-prename descclassname"> <span class="pre">&lt;queue&gt;</span></span><a class="headerlink" href="#cmdoption-swibrid-test-queue" title="Link to this definition"></a></dt>
<dd><p>slurm queue [medium]</p>
</dd></dl>

</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="output.html" class="btn btn-neutral float-left" title="understanding output" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Benedikt Obermayer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>