

<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>quick start guide &mdash; swibrid 2024 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js?v=172f29ee"></script>
      <script src="_static/doctools.js?v=888ff710"></script>
      <script src="_static/sphinx_highlight.js?v=4825356b"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="configuration" href="configure.html" />
    <link rel="prev" title="swibrid documentation" href="index.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            swibrid
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">quick start guide</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#installation">installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="#testing">testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="#running-your-own-data">running your own data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="configure.html">configuration</a></li>
<li class="toctree-l1"><a class="reference internal" href="output.html">understanding output</a></li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">command-line interface</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">swibrid</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">quick start guide</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/quickstart.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="quick-start-guide">
<h1>quick start guide<a class="headerlink" href="#quick-start-guide" title="Permalink to this heading"></a></h1>
<section id="installation">
<h2>installation<a class="headerlink" href="#installation" title="Permalink to this heading"></a></h2>
<ol class="arabic">
<li><p>clone the github repo and change into the source folder:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">git</span> <span class="n">clone</span> <span class="n">git</span><span class="nd">@github</span><span class="o">.</span><span class="n">com</span><span class="p">:</span><span class="n">bihealth</span><span class="o">/</span><span class="n">swibrid</span><span class="o">.</span><span class="n">git</span>
<span class="n">cd</span> <span class="n">swibrid</span>
</pre></div>
</div>
</li>
<li><p>create a conda environment:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">env</span> <span class="n">create</span> <span class="o">-</span><span class="n">f</span> <span class="n">swibrid_env</span><span class="o">.</span><span class="n">yaml</span>
<span class="n">conda</span> <span class="n">activate</span> <span class="n">swibrid_env</span>
</pre></div>
</div>
</li>
<li><p>install <code class="docutils literal notranslate"><span class="pre">swibrid</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="o">.</span>
</pre></div>
</div>
</li>
</ol>
</section>
<section id="testing">
<h2>testing<a class="headerlink" href="#testing" title="Permalink to this heading"></a></h2>
<p>for a simple and (relatively) quick end-to-end test, run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">swibrid</span> <span class="n">test</span>
</pre></div>
</div>
<p>this will create two samples with about 1000 synthetic reads in <code class="docutils literal notranslate"><span class="pre">input</span></code> and run the pipeline on this data,
using a reduced hg38 genome in <code class="docutils literal notranslate"><span class="pre">index</span></code> with only the switch region (chr14:105000000-106000000).
it will probably take about 10 minutes and produce plots in <code class="docutils literal notranslate"><span class="pre">output/read_plots</span></code> and
table of summary statistics in <code class="docutils literal notranslate"><span class="pre">output/summary</span></code></p>
</section>
<section id="running-your-own-data">
<h2>running your own data<a class="headerlink" href="#running-your-own-data" title="Permalink to this heading"></a></h2>
<p>this assumes you have a <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file with sequencing output from minION or PacBio.
If samples were multiplexed (e.g., with ONT barcodes), you should set up a sample sheet like so:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">BC01</span>         <span class="n">sample1</span>
<span class="n">BC02</span>         <span class="n">sample2</span>
<span class="o">...</span>
</pre></div>
</div>
<p>and a file with barcode and primer sequences like so:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span><span class="n">BC01</span>
<span class="n">AAGAAAGTTGTCGGTGTCTTTGTG</span>
<span class="o">&gt;</span><span class="n">BC02</span>
<span class="n">TCGATTCCGTTTGTAGTCGTCTGT</span>
<span class="o">...</span>
<span class="o">&gt;</span><span class="n">primer_mu_fw</span>
<span class="n">CACCCTTGAAAGTAGCCCATGCCTTCC</span>
<span class="o">&gt;</span><span class="n">primer_alpha_rv</span>
<span class="n">CTCAGTCCAACACCCACCACTCC</span>
<span class="o">&gt;</span><span class="n">primer_gamma_rv</span>
<span class="n">CTGCCTCCCAGTGTCCTGCATTACTTCTG</span>
</pre></div>
</div>
<p>if you don’t have multiplexed data, you should still run the demultiplexing for primer detection; simply set up a dummy sample sheet and the file with primers; all reads will end up in <code class="docutils literal notranslate"><span class="pre">undetermined.fastq.gz</span></code></p>
<ol class="arabic">
<li><p>set up snakemake and config files in a new directory:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">results</span>
<span class="n">cd</span> <span class="n">results</span>
<span class="n">swibrid</span> <span class="n">setup</span>
</pre></div>
</div>
</li>
<li><p>provide genome (+ index), annotation files in <code class="docutils literal notranslate"><span class="pre">index</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">mkdir</span> <span class="n">index</span>
<span class="n">cd</span> <span class="n">index</span>
<span class="c1"># get hg38 genome from UCSC (or elsewhere)</span>
<span class="n">wget</span> <span class="n">http</span><span class="p">:</span><span class="o">//</span><span class="n">hgdownload</span><span class="o">.</span><span class="n">soe</span><span class="o">.</span><span class="n">ucsc</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">goldenpath</span><span class="o">/</span><span class="n">hg38</span><span class="o">/</span><span class="n">bigZips</span><span class="o">/</span><span class="n">hg38</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>
<span class="n">gunzip</span> <span class="n">hg38</span><span class="o">.</span><span class="n">fa</span><span class="o">.</span><span class="n">gz</span>
<span class="c1"># create LAST index</span>
<span class="n">lastdb</span> <span class="n">hg38db</span> <span class="n">hg38</span><span class="o">.</span><span class="n">fa</span>
<span class="c1"># download gene annotation from ENCODE (or elsewhere)</span>
<span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ebi</span><span class="o">.</span><span class="n">ac</span><span class="o">.</span><span class="n">uk</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">databases</span><span class="o">/</span><span class="n">gencode</span><span class="o">/</span><span class="n">Gencode_human</span><span class="o">/</span><span class="n">release_33</span><span class="o">/</span><span class="n">gencode</span><span class="o">.</span><span class="n">v33</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">gunzip</span> <span class="n">gencode</span><span class="o">.</span><span class="n">v33</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">gz</span>
<span class="n">swibrid</span> <span class="n">get_annotation</span> <span class="o">-</span><span class="n">i</span> <span class="n">gencode</span><span class="o">.</span><span class="n">v33</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">gtf</span> <span class="o">-</span><span class="n">o</span> <span class="n">gencode</span><span class="o">.</span><span class="n">v33</span><span class="o">.</span><span class="n">annotation</span><span class="o">.</span><span class="n">exon</span><span class="o">.</span><span class="n">gene_shorted</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
</li>
<li><p>create bed file with switch region definitions:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chr14</span>   <span class="mi">105588700</span>       <span class="mi">105591700</span>       <span class="n">SA2</span>
<span class="n">chr14</span>   <span class="mi">105603000</span>       <span class="mi">105603500</span>       <span class="n">SE</span>
<span class="n">chr14</span>   <span class="mi">105626500</span>       <span class="mi">105629000</span>       <span class="n">SG4</span>
<span class="n">chr14</span>   <span class="mi">105645400</span>       <span class="mi">105647900</span>       <span class="n">SG2</span>
<span class="n">chr14</span>   <span class="mi">105708900</span>       <span class="mi">105712900</span>       <span class="n">SA1</span>
<span class="n">chr14</span>   <span class="mi">105743700</span>       <span class="mi">105747700</span>       <span class="n">SG1</span>
<span class="n">chr14</span>   <span class="mi">105772100</span>       <span class="mi">105775600</span>       <span class="n">SG3</span>
<span class="n">chr14</span>   <span class="mi">105856100</span>       <span class="mi">105861100</span>       <span class="n">SM</span>
</pre></div>
</div>
</li>
<li><p>edit (at least) the following entries in the <code class="docutils literal notranslate"><span class="pre">config.yaml</span></code> file (make sure that sample names in <code class="docutils literal notranslate"><span class="pre">SAMPLES</span></code> all appear in the sample sheet):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">INPUT</span><span class="p">:</span> <span class="s2">&quot;path/to/input.fastq.gz&quot;</span>
<span class="n">SAMPLE_SHEET</span><span class="p">:</span> <span class="s2">&quot;path/to/sample_sheet.csv&quot;</span>
<span class="n">BARCODES_PRIMERS</span><span class="p">:</span> <span class="s2">&quot;path/to/barcodes_primers.fa&quot;</span>
<span class="n">SAMPLES</span><span class="p">:</span> <span class="p">[</span><span class="s2">&quot;sample1&quot;</span><span class="p">,</span><span class="s2">&quot;sample2&quot;</span><span class="p">,</span> <span class="o">...</span><span class="p">]</span>
<span class="n">SWITCH_ANNOTATION</span><span class="p">:</span> <span class="s2">&quot;path/to/switch_regions.bed&quot;</span>
</pre></div>
</div>
</li>
<li><p>run the pipeline:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">swibrid</span> <span class="n">run</span> <span class="o">-</span><span class="n">np</span>        <span class="c1"># for a dry-run</span>
<span class="n">swibrid</span> <span class="n">run</span>            <span class="c1"># for an actual run</span>
<span class="n">swibrid</span> <span class="n">run</span> <span class="o">--</span><span class="n">slurm</span>    <span class="c1"># submit to slurm</span>
<span class="n">swibrid</span> <span class="n">run</span> <span class="o">--</span><span class="n">unlock</span>   <span class="c1"># unlock snakemake before restarting an interrupted/killed instance</span>
</pre></div>
</div>
</li>
</ol>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="index.html" class="btn btn-neutral float-left" title="swibrid documentation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="configure.html" class="btn btn-neutral float-right" title="configuration" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024, Benedikt Obermayer.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>